// @strapi/core/dist/services/mcp/prompt-registry.js
'use strict';

var McpCapabilityRegistry = require('./internal/McpCapabilityRegistry.js');
var createSafeCapabilityRegistration = require('./utils/createSafeCapabilityRegistration.js');
var wrapCapabilityHandlerForMetrics = require('./metrics/wrapCapabilityHandlerForMetrics.js');

// eslint-disable-next-line import/extensions
function _class_private_field_loose_base(receiver, privateKey) {
    if (!Object.prototype.hasOwnProperty.call(receiver, privateKey)) {
        throw new TypeError("attempted to use private field on non-instance");
    }
    return receiver;
}
var id = 0;
function _class_private_field_loose_key(name) {
    return "__private_" + id++ + "_" + name;
}
var _strapi = /*#__PURE__*/ _class_private_field_loose_key("_strapi");
class McpPromptRegistry extends McpCapabilityRegistry.McpCapabilityRegistryBase {
    bind(mcpServer) {
        const strapi = _class_private_field_loose_base(this, _strapi)[_strapi];
        super.register((definition)=>{
            const { name, title, description, argsSchema, createHandler } = definition;
            return createSafeCapabilityRegistration.createSafeCapabilityRegistration({
                strapi,
                capabilityType: 'Prompt',
                name,
                createHandler,
                createFallbackHandler (errorMessage) {
                    return async ()=>({
                            messages: [
                                {
                                    role: 'user',
                                    content: {
                                        type: 'text',
                                        text: `Prompt "${name}" failed to initialize: ${errorMessage}`
                                    }
                                }
                            ]
                        });
                },
                createErrorResult (error) {
                    return {
                        messages: [
                            {
                                role: 'user',
                                content: {
                                    type: 'text',
                                    text: `Prompt "${name}" execution failed: ${error.message}`
                                }
                            }
                        ]
                    };
                },
                registerWithSdk (safeHandler) {
                    const sdkHandler = wrapCapabilityHandlerForMetrics.wrapCapabilityHandlerForMetrics(strapi, 'prompt', name, definition.telemetry, safeHandler);
                    return mcpServer.registerPrompt(name, {
                        title,
                        description,
                        // @ts-expect-error - Internal handler type mismatch due to optional argsSchema
                        argsSchema
                    }, sdkHandler);
                }
            });
        });
    }
    constructor(ctx){
        super(ctx.definitions), Object.defineProperty(this, _strapi, {
            writable: true,
            value: void 0
        });
        _class_private_field_loose_base(this, _strapi)[_strapi] = ctx.strapi;
    }
}

exports.McpPromptRegistry = McpPromptRegistry;
