// @strapi/core/dist/services/mcp/resource-registry.mjs
import { McpCapabilityRegistryBase } from './internal/McpCapabilityRegistry.mjs';
import { createSafeCapabilityRegistration } from './utils/createSafeCapabilityRegistration.mjs';
import { wrapCapabilityHandlerForMetrics } from './metrics/wrapCapabilityHandlerForMetrics.mjs';

function _class_private_field_loose_base(receiver, privateKey) {
    if (!Object.prototype.hasOwnProperty.call(receiver, privateKey)) {
        throw new TypeError("attempted to use private field on non-instance");
    }
    return receiver;
}
var id = 0;
function _class_private_field_loose_key(name) {
    return "__private_" + id++ + "_" + name;
}
var _strapi = /*#__PURE__*/ _class_private_field_loose_key("_strapi");
class McpResourceRegistry extends McpCapabilityRegistryBase {
    bind(mcpServer) {
        const strapi = _class_private_field_loose_base(this, _strapi)[_strapi];
        super.register((definition)=>{
            const { name, uri, metadata, createHandler } = definition;
            return createSafeCapabilityRegistration({
                strapi,
                capabilityType: 'Resource',
                name,
                createHandler,
                createFallbackHandler (errorMessage) {
                    return async (resourceUri)=>({
                            contents: [
                                {
                                    uri: resourceUri.href,
                                    text: `Resource "${name}" failed to initialize: ${errorMessage}`,
                                    mimeType: 'text/plain'
                                }
                            ]
                        });
                },
                createErrorResult (error, args) {
                    return {
                        contents: [
                            {
                                uri: args[0] instanceof URL ? args[0].href : uri,
                                text: `Resource "${name}" execution failed: ${error.message}`,
                                mimeType: 'text/plain'
                            }
                        ]
                    };
                },
                registerWithSdk (safeHandler) {
                    const sdkHandler = wrapCapabilityHandlerForMetrics(strapi, 'resource', name, definition.telemetry, safeHandler);
                    return mcpServer.registerResource(name, uri, metadata, sdkHandler);
                }
            });
        });
    }
    constructor(ctx){
        super(ctx.definitions), Object.defineProperty(this, _strapi, {
            writable: true,
            value: void 0
        });
        _class_private_field_loose_base(this, _strapi)[_strapi] = ctx.strapi;
    }
}

export { McpResourceRegistry };
