// @strapi/content-manager/dist/server/mcp/derive-content-type-mcp-tools.js
'use strict';

var strapiUtils = require('@strapi/utils');
var permissionChecker = require('../services/permission-checker.js');
var utils = require('./utils.js');
var permissions = require('./permissions.js');
require('./schemas/blocks-schema.js');
var inputSchemas = require('./schemas/input-schemas.js');
var dataSchema = require('./schemas/data-schema.js');
var sortSchema = require('./schemas/sort-schema.js');
var filtersSchema = require('./schemas/filters-schema.js');
var outputSchemas = require('./schemas/output-schemas.js');
var collectionHandlers = require('./handlers/collection-handlers.js');
var singleTypeHandlers = require('./handlers/single-type-handlers.js');

// ---------------------------------------------------------------------------
// Collection-type tool-definition builder
// ---------------------------------------------------------------------------
/**
 * Derives the full set of `DerivedTool` definitions for a single collection-type model.
 * Produces list/get/create/update/delete tools, plus publish/unpublish/discard_draft
 * when draft-and-publish is enabled. Each tool's input/output schema is resolved
 * per-request so RBAC field and locale constraints are applied at call time.
 */ const buildCollectionTools = (strapi, model, ctx)=>{
    const uid = model.uid;
    const slug = utils.slugifyUidForMcpToolName(uid);
    const draftAndPublish = model.options?.draftAndPublish === true;
    const { attributes } = model;
    const runtimeLocaleSchema = permissions.buildLocaleSchema(ctx.localeCodes, ctx.defaultLocale);
    const resolveReadFields = (context)=>permissions.getPermittedFields(strapi, context.userAbility, permissionChecker.ACTIONS.read, uid, attributes);
    const resolveReadOutputSchema = (context)=>outputSchemas.buildDocumentOutputSchema(attributes, resolveReadFields(context));
    const resolveListInputSchema = (context)=>{
        const readFields = resolveReadFields(context);
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.read, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            locale: localeSchema,
            status: inputSchemas.statusSchema,
            page: inputSchemas.pageSchema,
            pageSize: inputSchemas.pageSizeSchema,
            sort: sortSchema.buildSortSchema(attributes, readFields),
            filters: filtersSchema.buildFiltersSchema(attributes, readFields)
        });
    };
    const resolveGetInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.read, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            documentId: inputSchemas.documentIdSchema,
            locale: localeSchema,
            status: inputSchemas.statusSchema
        });
    };
    const resolveCreateInputSchema = (context)=>{
        const writeFields = permissions.getPermittedFields(strapi, context.userAbility, permissionChecker.ACTIONS.create, uid, attributes);
        const dataSchema$1 = dataSchema.buildDataSchema(strapi, model, attributes, writeFields);
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.create, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            data: dataSchema$1,
            locale: localeSchema
        });
    };
    const resolveUpdateInputSchema = (context)=>{
        const writeFields = permissions.getPermittedFields(strapi, context.userAbility, permissionChecker.ACTIONS.update, uid, attributes);
        const dataSchema$1 = dataSchema.buildDataSchema(strapi, model, attributes, writeFields);
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.update, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            documentId: inputSchemas.documentIdSchema,
            data: dataSchema$1,
            locale: localeSchema
        });
    };
    const resolveDeleteInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.delete, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            documentId: inputSchemas.documentIdSchema,
            locale: localeSchema
        });
    };
    const resolvePublishInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.publish, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            documentId: inputSchemas.documentIdSchema,
            locale: localeSchema
        });
    };
    const resolveUnpublishInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.unpublish, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            documentId: inputSchemas.documentIdSchema,
            locale: localeSchema,
            discardDraft: strapiUtils.z.boolean().optional().describe('Also discard the draft when unpublishing.')
        });
    };
    const resolveDiscardDraftInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.discard, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            documentId: inputSchemas.documentIdSchema,
            locale: localeSchema
        });
    };
    const tools = [
        {
            name: `list_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'list'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'list'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.read,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveListInputSchema,
            resolveOutputSchema: (context)=>outputSchemas.buildListOutputSchema(attributes, resolveReadFields(context)),
            createHandler: collectionHandlers.createCollectionListHandler(uid)
        },
        {
            name: `get_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'get'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'get'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.read,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveGetInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: collectionHandlers.createCollectionGetHandler(uid)
        },
        {
            name: `create_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'create'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'create'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.create,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveCreateInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: collectionHandlers.createCollectionCreateHandler(uid)
        },
        {
            name: `update_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'update'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'update'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.update,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveUpdateInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: collectionHandlers.createCollectionUpdateHandler(uid)
        },
        {
            name: `delete_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'delete'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'delete'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.delete,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveDeleteInputSchema,
            resolveOutputSchema: (context)=>outputSchemas.buildDeleteOutputSchema(attributes, resolveReadFields(context)),
            createHandler: collectionHandlers.createCollectionDeleteHandler(uid)
        }
    ];
    if (draftAndPublish === true) {
        tools.push({
            name: `publish_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'publish'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'publish'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.publish,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolvePublishInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: collectionHandlers.createCollectionPublishHandler(uid)
        }, {
            name: `unpublish_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'unpublish'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'unpublish'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.unpublish,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveUnpublishInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: collectionHandlers.createCollectionUnpublishHandler(uid)
        }, {
            name: `discard_${slug}_draft`,
            telemetry: {
                source: 'content-manager',
                name: 'discard_draft'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'discard_draft'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.discard,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveDiscardDraftInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: collectionHandlers.createCollectionDiscardDraftHandler(uid)
        });
    }
    return tools;
};
// ---------------------------------------------------------------------------
// Single-type tool-definition builder
// ---------------------------------------------------------------------------
/**
 * Derives the full set of `DerivedTool` definitions for a single single-type model.
 * Produces get/write/delete tools, plus publish/unpublish/discard_draft when
 * draft-and-publish is enabled. Each tool's input/output schema is resolved
 * per-request so RBAC field and locale constraints are applied at call time.
 */ const buildSingleTypeTools = (strapi, model, ctx)=>{
    const uid = model.uid;
    const slug = utils.slugifyUidForMcpToolName(uid);
    const draftAndPublish = model.options?.draftAndPublish === true;
    const { attributes } = model;
    const runtimeLocaleSchema = permissions.buildLocaleSchema(ctx.localeCodes, ctx.defaultLocale);
    const resolveReadFields = (context)=>permissions.getPermittedFields(strapi, context.userAbility, permissionChecker.ACTIONS.read, uid, attributes);
    const resolveReadOutputSchema = (context)=>outputSchemas.buildDocumentOutputSchema(attributes, resolveReadFields(context));
    const resolveGetInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.read, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            locale: localeSchema,
            status: inputSchemas.statusSchema
        });
    };
    const resolveWriteInputSchema = (context)=>{
        const createFields = permissions.getPermittedFields(strapi, context.userAbility, permissionChecker.ACTIONS.create, uid, attributes);
        const updateFields = permissions.getPermittedFields(strapi, context.userAbility, permissionChecker.ACTIONS.update, uid, attributes);
        // null means all fields permitted; union of null with anything is null (all permitted)
        const writeFields = createFields === null || updateFields === null ? null : new Set([
            ...createFields,
            ...updateFields
        ]);
        const dataSchema$1 = dataSchema.buildDataSchema(strapi, model, attributes, writeFields);
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.update, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            data: dataSchema$1,
            locale: localeSchema
        });
    };
    const resolveDeleteInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.delete, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            locale: localeSchema
        });
    };
    const resolvePublishInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.publish, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            locale: localeSchema
        });
    };
    const resolveUnpublishInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.unpublish, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            locale: localeSchema,
            discardDraft: strapiUtils.z.boolean().optional().describe('Also discard the draft when unpublishing.')
        });
    };
    const resolveDiscardDraftInputSchema = (context)=>{
        const localeSchema = permissions.resolvePermittedLocaleSchema(strapi, context, permissionChecker.ACTIONS.discard, uid, ctx.localeCodes, ctx.defaultLocale, runtimeLocaleSchema);
        return strapiUtils.z.object({
            locale: localeSchema
        });
    };
    const tools = [
        {
            name: `get_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'get'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'get'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.read,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveGetInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: singleTypeHandlers.createSingleGetHandler(uid)
        },
        {
            name: `write_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'write'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'write'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.create,
                        subject: uid
                    },
                    {
                        action: permissionChecker.ACTIONS.update,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveWriteInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: singleTypeHandlers.createSingleWriteHandler(uid)
        },
        {
            name: `delete_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'delete'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'delete'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.delete,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveDeleteInputSchema,
            resolveOutputSchema: (context)=>outputSchemas.buildDeleteOutputSchema(attributes, resolveReadFields(context)),
            createHandler: singleTypeHandlers.createSingleDeleteHandler(uid)
        }
    ];
    if (draftAndPublish === true) {
        tools.push({
            name: `publish_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'publish'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'publish'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.publish,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolvePublishInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: singleTypeHandlers.createSinglePublishHandler(uid)
        }, {
            name: `unpublish_${slug}`,
            telemetry: {
                source: 'content-manager',
                name: 'unpublish'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'unpublish'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.unpublish,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveUnpublishInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: singleTypeHandlers.createSingleUnpublishHandler(uid)
        }, {
            name: `discard_${slug}_draft`,
            telemetry: {
                source: 'content-manager',
                name: 'discard_draft'
            },
            ...utils.describeTool({
                apiID: model.apiID,
                uid,
                operation: 'discard_draft'
            }),
            auth: {
                policies: [
                    {
                        action: permissionChecker.ACTIONS.discard,
                        subject: uid
                    }
                ]
            },
            resolveInputSchema: resolveDiscardDraftInputSchema,
            resolveOutputSchema: resolveReadOutputSchema,
            createHandler: singleTypeHandlers.createSingleDiscardDraftHandler(uid)
        });
    }
    return tools;
};
// ---------------------------------------------------------------------------
// Main export
// ---------------------------------------------------------------------------
/**
 * Builds MCP tool definitions for displayed content-manager models.
 * Visibility is enforced separately via static auth on each tool and MCP session capability sync.
 */ const deriveDisplayedContentTypeMcpToolDefinitions = (strapi, models, ctx = {
    localeCodes: null,
    defaultLocale: null
})=>{
    const tools = [];
    for (const model of models){
        if (model.kind === 'collectionType') {
            tools.push(...buildCollectionTools(strapi, model, ctx));
        } else if (model.kind === 'singleType') {
            tools.push(...buildSingleTypeTools(strapi, model, ctx));
        }
    }
    return tools;
};

exports.slugifyUidForMcpToolName = utils.slugifyUidForMcpToolName;
exports.getComponentLeafPaths = permissions.getComponentLeafPaths;
exports.buildDataSchema = dataSchema.buildDataSchema;
exports.buildSortSchema = sortSchema.buildSortSchema;
exports.buildFiltersSchema = filtersSchema.buildFiltersSchema;
exports.deriveDisplayedContentTypeMcpToolDefinitions = deriveDisplayedContentTypeMcpToolDefinitions;
