// @strapi/review-workflows/dist/admin/routes/content-manager/model/constants.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var AssigneeFilter = require('./components/AssigneeFilter.js');
var StageFilter = require('./components/StageFilter.js');
var TableColumns = require('./components/TableColumns.js');
var constants = require('./id/components/constants.js');

const REVIEW_WORKFLOW_COLUMNS = [
    {
        name: constants.STAGE_ATTRIBUTE_NAME,
        attribute: {
            type: 'relation',
            relation: 'oneToMany',
            target: 'admin::review-workflow-stage'
        },
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.stage',
            defaultMessage: 'review stage'
        },
        searchable: false,
        sortable: true,
        mainField: {
            name: 'name',
            type: 'string'
        },
        cellFormatter: (props)=>/*#__PURE__*/ jsxRuntime.jsx(TableColumns.StageColumn, {
                ...props
            })
    },
    {
        name: constants.ASSIGNEE_ATTRIBUTE_NAME,
        attribute: {
            type: 'relation',
            target: 'admin::user',
            relation: 'oneToMany'
        },
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.assignee',
            defaultMessage: 'assignee'
        },
        searchable: false,
        sortable: true,
        mainField: {
            name: 'firstname',
            type: 'string'
        },
        cellFormatter: (props)=>/*#__PURE__*/ jsxRuntime.jsx(TableColumns.AssigneeColumn, {
                ...props
            })
    }
];
const REVIEW_WORKFLOW_FILTERS = [
    {
        mainField: {
            name: 'name',
            type: 'string'
        },
        input: StageFilter.StageFilter,
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.stage',
            defaultMessage: 'review stage'
        },
        name: 'strapi_stage',
        type: 'relation'
    },
    {
        type: 'relation',
        mainField: {
            name: 'id',
            type: 'integer'
        },
        input: AssigneeFilter.AssigneeFilter,
        operators: [
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$eq',
                    defaultMessage: 'is'
                },
                value: '$eq'
            },
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$ne',
                    defaultMessage: 'is not'
                },
                value: '$ne'
            },
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$null',
                    defaultMessage: 'is null'
                },
                value: '$null'
            },
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$notNull',
                    defaultMessage: 'is not null'
                },
                value: '$notNull'
            }
        ],
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.assignee',
            defaultMessage: 'assignee'
        },
        name: 'strapi_assignee'
    }
];

exports.REVIEW_WORKFLOW_COLUMNS = REVIEW_WORKFLOW_COLUMNS;
exports.REVIEW_WORKFLOW_FILTERS = REVIEW_WORKFLOW_FILTERS;
