// @strapi/review-workflows/dist/admin/routes/content-manager/model/constants.mjs
import { jsx } from 'react/jsx-runtime';
import { AssigneeFilter } from './components/AssigneeFilter.mjs';
import { StageFilter } from './components/StageFilter.mjs';
import { StageColumn, AssigneeColumn } from './components/TableColumns.mjs';
import { STAGE_ATTRIBUTE_NAME, ASSIGNEE_ATTRIBUTE_NAME } from './id/components/constants.mjs';

const REVIEW_WORKFLOW_COLUMNS = [
    {
        name: STAGE_ATTRIBUTE_NAME,
        attribute: {
            type: 'relation',
            relation: 'oneToMany',
            target: 'admin::review-workflow-stage'
        },
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.stage',
            defaultMessage: 'review stage'
        },
        searchable: false,
        sortable: true,
        mainField: {
            name: 'name',
            type: 'string'
        },
        cellFormatter: (props)=>/*#__PURE__*/ jsx(StageColumn, {
                ...props
            })
    },
    {
        name: ASSIGNEE_ATTRIBUTE_NAME,
        attribute: {
            type: 'relation',
            target: 'admin::user',
            relation: 'oneToMany'
        },
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.assignee',
            defaultMessage: 'assignee'
        },
        searchable: false,
        sortable: true,
        mainField: {
            name: 'firstname',
            type: 'string'
        },
        cellFormatter: (props)=>/*#__PURE__*/ jsx(AssigneeColumn, {
                ...props
            })
    }
];
const REVIEW_WORKFLOW_FILTERS = [
    {
        mainField: {
            name: 'name',
            type: 'string'
        },
        input: StageFilter,
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.stage',
            defaultMessage: 'review stage'
        },
        name: 'strapi_stage',
        type: 'relation'
    },
    {
        type: 'relation',
        mainField: {
            name: 'id',
            type: 'integer'
        },
        input: AssigneeFilter,
        operators: [
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$eq',
                    defaultMessage: 'is'
                },
                value: '$eq'
            },
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$ne',
                    defaultMessage: 'is not'
                },
                value: '$ne'
            },
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$null',
                    defaultMessage: 'is null'
                },
                value: '$null'
            },
            {
                label: {
                    id: 'components.FilterOptions.FILTER_TYPES.$notNull',
                    defaultMessage: 'is not null'
                },
                value: '$notNull'
            }
        ],
        label: {
            id: 'review-workflows.containers.list.table-headers.reviewWorkflows.assignee',
            defaultMessage: 'assignee'
        },
        name: 'strapi_assignee'
    }
];

export { REVIEW_WORKFLOW_COLUMNS, REVIEW_WORKFLOW_FILTERS };
