// @strapi/core/dist/services/mcp/internal/McpServerFactory.js
'use strict';

var mcp_js = require('@modelcontextprotocol/sdk/server/mcp.js');
var promptRegistry = require('../prompt-registry.js');
var resourceRegistry = require('../resource-registry.js');
var toolRegistry = require('../tool-registry.js');
var syncMcpSessionCapabilities = require('./syncMcpSessionCapabilities.js');

// eslint-disable-next-line import/extensions
const createMcpServerWithRegistries = ({ strapi, definitions, isDevMode, ability, user })=>{
    const capabilities = {
        logging: {}
    };
    // Advertise capability categories when definitions exist (server-level), not per-user enabled
    // count. Clients discover the real set via tools/list, prompts/list, resources/list after sync.
    if (definitions.tools.size > 0) {
        capabilities.tools = {};
    }
    if (definitions.prompts.size > 0) {
        capabilities.prompts = {};
    }
    if (definitions.resources.size > 0) {
        capabilities.resources = {};
    }
    const mcpServer = new mcp_js.McpServer({
        name: 'strapi-mcp-server',
        version: '1.0.0'
    }, {
        capabilities
    });
    // Bootstrap registries with current definitions
    const tools = new toolRegistry.McpToolRegistry({
        strapi,
        definitions: definitions.tools,
        ability,
        user
    });
    const prompts = new promptRegistry.McpPromptRegistry({
        strapi,
        definitions: definitions.prompts
    });
    const resources = new resourceRegistry.McpResourceRegistry({
        strapi,
        definitions: definitions.resources
    });
    // Register capabilities (disabled by default)
    tools.bind(mcpServer);
    prompts.bind(mcpServer);
    resources.bind(mcpServer);
    syncMcpSessionCapabilities.syncMcpSessionCapabilities({
        registries: {
            tools,
            prompts,
            resources
        },
        definitions,
        ability,
        isDevMode
    });
    return {
        mcpServer,
        registries: {
            tools,
            prompts,
            resources
        }
    };
};

exports.createMcpServerWithRegistries = createMcpServerWithRegistries;
