// @strapi/admin/dist/admin/admin/src/pages/Settings/pages/Roles/components/ConditionsModal.mjs
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import * as React from 'react';
import { Modal, Breadcrumbs, Crumb, Box, Typography, Button, Flex, MultiSelectNested } from '@strapi/design-system';
import { produce } from 'immer';
import get from 'lodash/get';
import groupBy from 'lodash/groupBy';
import upperFirst from 'lodash/upperFirst';
import { useIntl } from 'react-intl';
import { capitalise } from '../../../../../utils/strings.mjs';
import { usePermissionsDataManager } from '../hooks/usePermissionsDataManager.mjs';

const ConditionsModal = ({ actions = [], headerBreadCrumbs = [], isFormDisabled, isReadOnly = false, onClose })=>{
    const { formatMessage } = useIntl();
    const { availableConditions, modifiedData, onChangeConditions } = usePermissionsDataManager();
    const arrayOfOptionsGroupedByCategory = React.useMemo(()=>{
        return Object.entries(groupBy(availableConditions, 'category'));
    }, [
        availableConditions
    ]);
    const actionsToDisplay = React.useMemo(()=>actions.filter(// @ts-expect-error – TODO: fix this type issue
        ({ isDisplayed, hasSomeActionsSelected, hasAllActionsSelected })=>isDisplayed && Boolean(hasSomeActionsSelected || hasAllActionsSelected)), [
        actions
    ]);
    const [state, setState] = React.useState(()=>createDefaultConditionsForm(actionsToDisplay, modifiedData, arrayOfOptionsGroupedByCategory));
    React.useEffect(()=>{
        if (isReadOnly === false) {
            return;
        }
        setState(createDefaultConditionsForm(actionsToDisplay, modifiedData, arrayOfOptionsGroupedByCategory));
    }, [
        isReadOnly,
        actionsToDisplay,
        modifiedData,
        arrayOfOptionsGroupedByCategory
    ]);
    const handleChange = React.useCallback((name, values)=>{
        if (isReadOnly === true) {
            return;
        }
        setState(produce((draft)=>{
            if (draft[name] === undefined) {
                draft[name] = {};
            }
            if (draft[name].default === undefined) {
                draft[name].default = {};
            }
            draft[name].default = values;
        }));
    }, [
        isReadOnly
    ]);
    const handleSubmit = React.useCallback(()=>{
        if (isReadOnly === true) {
            onClose?.();
            return;
        }
        const conditionsWithoutCategory = Object.entries(state).reduce((acc, current)=>{
            const [key, value] = current;
            const merged = Object.values(value).reduce((acc1, current1)=>{
                return {
                    ...acc1,
                    ...current1
                };
            }, {});
            acc[key] = merged;
            return acc;
        }, {});
        onChangeConditions(conditionsWithoutCategory);
        onClose?.();
    }, [
        isReadOnly,
        state,
        onChangeConditions,
        onClose
    ]);
    const onCloseModal = React.useCallback(()=>{
        setState(createDefaultConditionsForm(actionsToDisplay, modifiedData, arrayOfOptionsGroupedByCategory));
        onClose?.();
    }, [
        actionsToDisplay,
        modifiedData,
        arrayOfOptionsGroupedByCategory,
        onClose
    ]);
    return /*#__PURE__*/ jsxs(Modal.Content, {
        children: [
            /*#__PURE__*/ jsx(Modal.Header, {
                children: /*#__PURE__*/ jsx(Breadcrumbs, {
                    id: "condition-modal-breadcrumbs",
                    label: headerBreadCrumbs.join(', '),
                    children: headerBreadCrumbs.map((label, index, arr)=>/*#__PURE__*/ jsx(Crumb, {
                            isCurrent: index === arr.length - 1,
                            children: upperFirst(formatMessage({
                                id: label,
                                defaultMessage: label
                            }))
                        }, label))
                })
            }),
            /*#__PURE__*/ jsxs(Modal.Body, {
                children: [
                    isReadOnly && /*#__PURE__*/ jsx(Box, {
                        paddingBottom: 4,
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: formatMessage({
                                id: 'Settings.permissions.conditions.inherited-readonly',
                                defaultMessage: 'These conditions are inherited from your permissions and are read-only.'
                            })
                        })
                    }),
                    actionsToDisplay.length === 0 && /*#__PURE__*/ jsx(Typography, {
                        children: formatMessage({
                            id: 'Settings.permissions.conditions.no-actions',
                            defaultMessage: 'You first need to select actions (create, read, update, ...) before defining conditions on them.'
                        })
                    }),
                    /*#__PURE__*/ jsx("ul", {
                        children: actionsToDisplay.map(({ actionId, label, pathToConditionsObject }, index)=>{
                            const name = pathToConditionsObject.join('..');
                            return /*#__PURE__*/ jsx(ActionRow, {
                                arrayOfOptionsGroupedByCategory: arrayOfOptionsGroupedByCategory,
                                label: label,
                                isFormDisabled: isFormDisabled || isReadOnly,
                                isReadOnly: isReadOnly,
                                isGrey: index % 2 === 0,
                                name: name,
                                onChange: handleChange,
                                value: get(state, name, {})
                            }, actionId);
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsx(Modal.Footer, {
                children: isReadOnly ? /*#__PURE__*/ jsx(Button, {
                    variant: "tertiary",
                    onClick: ()=>onCloseModal(),
                    children: formatMessage({
                        id: 'app.components.Button.close',
                        defaultMessage: 'Close'
                    })
                }) : /*#__PURE__*/ jsxs(Fragment, {
                    children: [
                        /*#__PURE__*/ jsx(Button, {
                            variant: "tertiary",
                            onClick: ()=>onCloseModal(),
                            children: formatMessage({
                                id: 'app.components.Button.cancel',
                                defaultMessage: 'Cancel'
                            })
                        }),
                        /*#__PURE__*/ jsx(Button, {
                            onClick: handleSubmit,
                            children: formatMessage({
                                id: 'Settings.permissions.conditions.apply',
                                defaultMessage: 'Apply'
                            })
                        })
                    ]
                })
            })
        ]
    });
};
const createDefaultConditionsForm = (actionsToDisplay, modifiedData, arrayOfOptionsGroupedByCategory)=>{
    return actionsToDisplay.reduce((acc, current)=>{
        const valueFromModifiedData = get(modifiedData, [
            ...current.pathToConditionsObject,
            'conditions'
        ], {});
        const categoryDefaultForm = arrayOfOptionsGroupedByCategory.reduce((acc, current)=>{
            const [categoryName, relatedConditions] = current;
            const conditionsForm = relatedConditions.reduce((acc, current)=>{
                acc[current.id] = get(valueFromModifiedData, current.id, false);
                return acc;
            }, {});
            acc[categoryName] = conditionsForm;
            return acc;
        }, {});
        acc[current.pathToConditionsObject.join('..')] = categoryDefaultForm;
        return acc;
    }, {});
};
const ActionRow = ({ arrayOfOptionsGroupedByCategory, isFormDisabled = false, isGrey = false, isReadOnly = false, label, name, onChange, value })=>{
    const { formatMessage } = useIntl();
    const handleChange = (val)=>{
        if (onChange) {
            onChange(name, getNewStateFromChangedValues(arrayOfOptionsGroupedByCategory, val));
        }
    };
    const selectedValues = getSelectedValues(value);
    return /*#__PURE__*/ jsxs(Flex, {
        tag: "li",
        background: isGrey ? 'neutral100' : 'neutral0',
        paddingBottom: 3,
        paddingTop: 3,
        justifyContent: 'space-evenly',
        children: [
            /*#__PURE__*/ jsxs(Flex, {
                style: {
                    width: 180
                },
                children: [
                    /*#__PURE__*/ jsxs(Typography, {
                        variant: "sigma",
                        textColor: "neutral600",
                        children: [
                            formatMessage({
                                id: 'Settings.permissions.conditions.can',
                                defaultMessage: 'Can'
                            }),
                            " "
                        ]
                    }),
                    /*#__PURE__*/ jsx(Typography, {
                        variant: "sigma",
                        title: label,
                        textColor: "primary600",
                        ellipsis: true,
                        children: formatMessage({
                            id: `Settings.roles.form.permissions.${label.toLowerCase()}`,
                            defaultMessage: label
                        })
                    }),
                    /*#__PURE__*/ jsxs(Typography, {
                        variant: "sigma",
                        textColor: "neutral600",
                        children: [
                            " ",
                            formatMessage({
                                id: 'Settings.permissions.conditions.when',
                                defaultMessage: 'When'
                            })
                        ]
                    })
                ]
            }),
            /*#__PURE__*/ jsx(Box, {
                style: {
                    maxWidth: 430,
                    width: '100%'
                },
                children: isReadOnly ? /*#__PURE__*/ jsx(Flex, {
                    direction: "column",
                    alignItems: "flex-start",
                    gap: 3,
                    children: selectedValues.length === 0 ? /*#__PURE__*/ jsx(Typography, {
                        textColor: "neutral600",
                        children: formatMessage({
                            id: 'Settings.permissions.conditions.none-applied',
                            defaultMessage: 'No conditions applied.'
                        })
                    }) : arrayOfOptionsGroupedByCategory.map(([categoryName, conditions])=>{
                        const selectedConditions = conditions.filter((c)=>selectedValues.includes(c.id));
                        if (selectedConditions.length === 0) {
                            return null;
                        }
                        return /*#__PURE__*/ jsxs(Flex, {
                            direction: "column",
                            alignItems: "flex-start",
                            gap: 2,
                            children: [
                                /*#__PURE__*/ jsx(Typography, {
                                    variant: "pi",
                                    fontWeight: "bold",
                                    textColor: "neutral600",
                                    children: categoryName
                                }),
                                /*#__PURE__*/ jsx(Box, {
                                    children: /*#__PURE__*/ jsx(Typography, {
                                        variant: "omega",
                                        textColor: "neutral800",
                                        children: selectedConditions.map((condition)=>condition.displayName).join(', ')
                                    })
                                })
                            ]
                        }, categoryName);
                    }).filter(Boolean)
                }) : /*#__PURE__*/ jsx(MultiSelectNested, {
                    id: name,
                    customizeContent: (values = [])=>`${values.length} currently selected`,
                    onChange: handleChange,
                    value: selectedValues,
                    options: getNestedOptions(arrayOfOptionsGroupedByCategory),
                    disabled: isFormDisabled
                })
            })
        ]
    });
};
const getSelectedValues = (rawValue)=>Object.values(rawValue).map((x)=>Object.entries(x).filter(([, value])=>value).map(([key])=>key)).flat();
const getNestedOptions = (options)=>options.reduce((acc, [label, children])=>{
        acc.push({
            label: capitalise(label),
            children: children.map((child)=>({
                    label: child.displayName,
                    value: child.id
                }))
        });
        return acc;
    }, []);
const getNewStateFromChangedValues = (options, changedValues)=>options.map(([, values])=>values).flat().reduce((acc, curr)=>({
            [curr.id]: changedValues.includes(curr.id),
            ...acc
        }), {});

export { ConditionsModal };
