// @strapi/admin/dist/admin/ee/admin/src/services/ai.js
'use strict';

var api = require('../../../../admin/src/services/api.js');

const aiService = api.adminApi.injectEndpoints({
    endpoints: (builder)=>({
            getAiUsage: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: `/admin/ai-usage`
                    }),
                providesTags: [
                    'AiUsage'
                ]
            }),
            getAiToken: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: '/admin/ai-token'
                    }),
                transformResponse (res) {
                    return res.data;
                }
            }),
            getAiFeatureConfig: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: '/admin/ai-feature-config'
                    }),
                transformResponse (res) {
                    return res.data;
                },
                providesTags: [
                    'AiFeatureConfig'
                ]
            })
        }),
    overrideExisting: true
});
const { useGetAiUsageQuery, useGetAiTokenQuery, useLazyGetAiTokenQuery, useGetAiFeatureConfigQuery } = aiService;

exports.useGetAiFeatureConfigQuery = useGetAiFeatureConfigQuery;
exports.useGetAiTokenQuery = useGetAiTokenQuery;
exports.useGetAiUsageQuery = useGetAiUsageQuery;
exports.useLazyGetAiTokenQuery = useLazyGetAiTokenQuery;
