// @strapi/admin/dist/admin/ee/admin/src/services/ai.mjs
import { adminApi } from '../../../../admin/src/services/api.mjs';

const aiService = adminApi.injectEndpoints({
    endpoints: (builder)=>({
            getAiUsage: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: `/admin/ai-usage`
                    }),
                providesTags: [
                    'AiUsage'
                ]
            }),
            getAiToken: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: '/admin/ai-token'
                    }),
                transformResponse (res) {
                    return res.data;
                }
            }),
            getAiFeatureConfig: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: '/admin/ai-feature-config'
                    }),
                transformResponse (res) {
                    return res.data;
                },
                providesTags: [
                    'AiFeatureConfig'
                ]
            })
        }),
    overrideExisting: true
});
const { useGetAiUsageQuery, useGetAiTokenQuery, useLazyGetAiTokenQuery, useGetAiFeatureConfigQuery } = aiService;

export { useGetAiFeatureConfigQuery, useGetAiTokenQuery, useGetAiUsageQuery, useLazyGetAiTokenQuery };
