// @strapi/admin/dist/admin/admin/src/pages/Settings/pages/AdminTokens/EditView/components/AdminPermissions.mjs
import { jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { Page } from '../../../../../../components/PageHelpers.mjs';
import { useAuth } from '../../../../../../features/Auth.mjs';
import { useGetAPITokenOwnerPermissionsQuery } from '../../../../../../services/apiTokens.mjs';
import { useGetRolePermissionLayoutQuery } from '../../../../../../services/users.mjs';
import { Permissions } from '../../../Roles/components/Permissions.mjs';

const AdminPermissions = /*#__PURE__*/ React.forwardRef(({ disabled, initialAdminPermissions, tokenId, ownerUserId }, ref)=>{
    const { permissions: currentUserPermissions, user: currentUser } = useAuth('AdminPermissions', (auth)=>auth);
    const isCurrentUserOwner = ownerUserId === undefined || ownerUserId === null || ownerUserId === currentUser?.id;
    const { data: ownerPermissions } = useGetAPITokenOwnerPermissionsQuery(tokenId, {
        skip: isCurrentUserOwner || tokenId === undefined,
        refetchOnMountOrArgChange: true
    });
    const { data: layout, isLoading, error } = useGetRolePermissionLayoutQuery({
        role: ''
    });
    const permissionsFormKey = React.useMemo(()=>JSON.stringify({
            tokenId: tokenId ?? 'create',
            ownerUserId: ownerUserId ?? 'current-user',
            initialAdminPermissions
        }), [
        initialAdminPermissions,
        ownerUserId,
        tokenId
    ]);
    if (isLoading) {
        return /*#__PURE__*/ jsx(Page.Loading, {});
    }
    if (error !== undefined || layout === undefined) {
        return null;
    }
    // If the current user is not the owner, but we failed to fetch the owner's permissions, we can't display the permissions form.
    if (isCurrentUserOwner === false && ownerPermissions === undefined) {
        return null;
    }
    const effectivePermissions = isCurrentUserOwner ? currentUserPermissions : ownerPermissions;
    return /*#__PURE__*/ jsx(Permissions, {
        ref: ref,
        layout: layout,
        permissions: initialAdminPermissions,
        userPermissions: effectivePermissions,
        isFormDisabled: disabled
    }, permissionsFormKey);
});

export { AdminPermissions };
