// @strapi/admin/dist/admin/admin/src/pages/Settings/pages/AdminTokens/EditView/components/AdminPermissions.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var PageHelpers = require('../../../../../../components/PageHelpers.js');
var Auth = require('../../../../../../features/Auth.js');
var apiTokens = require('../../../../../../services/apiTokens.js');
var users = require('../../../../../../services/users.js');
var Permissions = require('../../../Roles/components/Permissions.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const AdminPermissions = /*#__PURE__*/ React__namespace.forwardRef(({ disabled, initialAdminPermissions, tokenId, ownerUserId }, ref)=>{
    const { permissions: currentUserPermissions, user: currentUser } = Auth.useAuth('AdminPermissions', (auth)=>auth);
    const isCurrentUserOwner = ownerUserId === undefined || ownerUserId === null || ownerUserId === currentUser?.id;
    const { data: ownerPermissions } = apiTokens.useGetAPITokenOwnerPermissionsQuery(tokenId, {
        skip: isCurrentUserOwner || tokenId === undefined,
        refetchOnMountOrArgChange: true
    });
    const { data: layout, isLoading, error } = users.useGetRolePermissionLayoutQuery({
        role: ''
    });
    const permissionsFormKey = React__namespace.useMemo(()=>JSON.stringify({
            tokenId: tokenId ?? 'create',
            ownerUserId: ownerUserId ?? 'current-user',
            initialAdminPermissions
        }), [
        initialAdminPermissions,
        ownerUserId,
        tokenId
    ]);
    if (isLoading) {
        return /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Loading, {});
    }
    if (error !== undefined || layout === undefined) {
        return null;
    }
    // If the current user is not the owner, but we failed to fetch the owner's permissions, we can't display the permissions form.
    if (isCurrentUserOwner === false && ownerPermissions === undefined) {
        return null;
    }
    const effectivePermissions = isCurrentUserOwner ? currentUserPermissions : ownerPermissions;
    return /*#__PURE__*/ jsxRuntime.jsx(Permissions.Permissions, {
        ref: ref,
        layout: layout,
        permissions: initialAdminPermissions,
        userPermissions: effectivePermissions,
        isFormDisabled: disabled
    }, permissionsFormKey);
});

exports.AdminPermissions = AdminPermissions;
