// @strapi/admin/dist/server/server/src/validation/admin-tokens.js
'use strict';

var utils = require('@strapi/utils');
var constants = require('../services/constants.js');
var commonValidators = require('./common-validators.js');

const adminTokenCreationSchema = utils.yup.object().shape({
    kind: utils.yup.string().oneOf([
        'admin'
    ]).optional(),
    name: utils.yup.string().min(1).required(),
    description: utils.yup.string().optional(),
    lifespan: utils.yup.number().min(1).oneOf(Object.values(constants.API_TOKEN_LIFESPANS)).nullable(),
    adminPermissions: utils.yup.array().of(commonValidators.permission),
    // adminUserOwner is set by the controller from ctx.state.user (full user object) or a strapiID from body
    adminUserOwner: utils.yup.mixed().nullable()
}).noUnknown().strict();
const adminTokenUpdateSchema = utils.yup.object().shape({
    name: utils.yup.string().min(1).notNull(),
    description: utils.yup.string().nullable(),
    adminPermissions: utils.yup.array().of(commonValidators.permission).nullable()
}).noUnknown().strict();
const validateAdminTokenCreationInput = utils.validateYupSchema(adminTokenCreationSchema);
const validateAdminTokenUpdateInput = utils.validateYupSchema(adminTokenUpdateSchema);

exports.validateAdminTokenCreationInput = validateAdminTokenCreationInput;
exports.validateAdminTokenUpdateInput = validateAdminTokenUpdateInput;
