// @strapi/admin/dist/server/server/src/validation/project-settings.js
'use strict';

var utils = require('@strapi/utils');

const MAX_IMAGE_WIDTH = 750;
const MAX_IMAGE_HEIGHT = MAX_IMAGE_WIDTH;
const MAX_IMAGE_FILE_SIZE = 1024 * 1024; // 1Mo
const updateProjectSettings = utils.z.object({
    menuLogo: utils.z.string().nullish(),
    authLogo: utils.z.string().nullish()
}).strict();
const updateProjectSettingsLogo = utils.z.object({
    originalFilename: utils.z.string().nullish(),
    mimetype: utils.z.enum([
        'image/jpeg',
        'image/png',
        'image/svg+xml'
    ]),
    size: utils.z.number().max(MAX_IMAGE_FILE_SIZE).nullish()
});
const updateProjectSettingsFiles = utils.z.object({
    menuLogo: updateProjectSettingsLogo.nullish(),
    authLogo: updateProjectSettingsLogo.nullish()
}).strict();
const logoDimensions = utils.z.object({
    width: utils.z.number().max(MAX_IMAGE_WIDTH).nullish(),
    height: utils.z.number().max(MAX_IMAGE_HEIGHT).nullish()
});
const updateProjectSettingsImagesDimensions = utils.z.object({
    menuLogo: logoDimensions.nullish(),
    authLogo: logoDimensions.nullish()
}).strict();
const validateUpdateProjectSettings = utils.validateZodSchema(updateProjectSettings);
const validateUpdateProjectSettingsFiles = utils.validateZodSchema(updateProjectSettingsFiles);
const validateUpdateProjectSettingsImagesDimensions = utils.validateZodSchema(updateProjectSettingsImagesDimensions);

exports.validateUpdateProjectSettings = validateUpdateProjectSettings;
exports.validateUpdateProjectSettingsFiles = validateUpdateProjectSettingsFiles;
exports.validateUpdateProjectSettingsImagesDimensions = validateUpdateProjectSettingsImagesDimensions;
