// @strapi/content-manager/dist/server/mcp/schemas/blocks-schema.js
'use strict';

var strapiUtils = require('@strapi/utils');

// ---------------------------------------------------------------------------
// Inline nodes
// ---------------------------------------------------------------------------
const textNodeSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('text'),
    text: strapiUtils.z.string().describe('The text content.'),
    bold: strapiUtils.z.boolean().optional(),
    italic: strapiUtils.z.boolean().optional(),
    underline: strapiUtils.z.boolean().optional(),
    strikethrough: strapiUtils.z.boolean().optional(),
    code: strapiUtils.z.boolean().optional()
}).describe('A text node with optional formatting marks.');
const linkNodeSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('link'),
    url: strapiUtils.z.string().describe('The URL the link points to.'),
    children: strapiUtils.z.array(textNodeSchema).min(1)
}).describe('An inline link node. Children must be text nodes.');
const inlineNodeSchema = strapiUtils.z.discriminatedUnion('type', [
    textNodeSchema,
    linkNodeSchema
]);
// ---------------------------------------------------------------------------
// Root block nodes
// ---------------------------------------------------------------------------
const paragraphSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('paragraph'),
    children: strapiUtils.z.array(inlineNodeSchema).min(1)
});
const headingSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('heading'),
    level: strapiUtils.z.union([
        strapiUtils.z.literal(1),
        strapiUtils.z.literal(2),
        strapiUtils.z.literal(3),
        strapiUtils.z.literal(4),
        strapiUtils.z.literal(5),
        strapiUtils.z.literal(6)
    ]),
    children: strapiUtils.z.array(inlineNodeSchema).min(1)
});
const quoteSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('quote'),
    children: strapiUtils.z.array(inlineNodeSchema).min(1)
});
const codeSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('code'),
    language: strapiUtils.z.string().nullable().optional().describe('Programming language identifier (e.g. "javascript", "python").'),
    children: strapiUtils.z.array(textNodeSchema).min(1)
});
// ---------------------------------------------------------------------------
// List nodes (recursive via z.lazy — mirrors yup.lazy in blocks-validator.ts)
// ---------------------------------------------------------------------------
const listItemSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('list-item'),
    children: strapiUtils.z.array(inlineNodeSchema).min(1)
});
// ZodType annotation is required because of the self-reference via z.lazy
const listSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('list'),
    format: strapiUtils.z.enum([
        'ordered',
        'unordered'
    ]),
    indentLevel: strapiUtils.z.number().optional(),
    children: strapiUtils.z.array(strapiUtils.z.lazy(()=>strapiUtils.z.union([
            listItemSchema,
            listSchema
        ]))).min(1).describe('Children must be list-item or nested list nodes.')
});
// ---------------------------------------------------------------------------
// Image block
// ---------------------------------------------------------------------------
const imageSchema = strapiUtils.z.object({
    type: strapiUtils.z.literal('image'),
    image: strapiUtils.z.object({
        name: strapiUtils.z.string(),
        alternativeText: strapiUtils.z.string().nullable().optional(),
        url: strapiUtils.z.string(),
        caption: strapiUtils.z.string().nullable().optional(),
        width: strapiUtils.z.number(),
        height: strapiUtils.z.number(),
        formats: strapiUtils.z.record(strapiUtils.z.string(), strapiUtils.z.unknown()).nullable().optional(),
        hash: strapiUtils.z.string(),
        ext: strapiUtils.z.string(),
        mime: strapiUtils.z.string(),
        size: strapiUtils.z.number(),
        previewUrl: strapiUtils.z.string().nullable().optional(),
        provider: strapiUtils.z.string(),
        provider_metadata: strapiUtils.z.unknown().nullable().optional(),
        createdAt: strapiUtils.z.string(),
        updatedAt: strapiUtils.z.string()
    }).describe('An existing media asset. Use media tools to retrieve this data — MCP cannot upload files.'),
    children: strapiUtils.z.array(strapiUtils.z.object({
        type: strapiUtils.z.literal('text'),
        text: strapiUtils.z.literal('')
    })).length(1)
});
// ---------------------------------------------------------------------------
// Top-level schema
// ---------------------------------------------------------------------------
const blockNodeSchema = strapiUtils.z.discriminatedUnion('type', [
    paragraphSchema,
    headingSchema,
    quoteSchema,
    codeSchema,
    imageSchema
]);
/**
 * Returns the Zod schema for a Strapi Blocks (rich-text) field input.
 * Accepts an array of block nodes — paragraph, heading, quote, code, image, and list —
 * matching Strapi's internal blocks data model.
 */ const buildBlocksInputSchema = ()=>strapiUtils.z.array(// listSchema uses z.lazy so it cannot participate in discriminatedUnion;
    // use a regular union to merge it back in.
    strapiUtils.z.union([
        blockNodeSchema,
        listSchema
    ])).describe('An array of block nodes representing structured rich text content.');

exports.buildBlocksInputSchema = buildBlocksInputSchema;
