// @strapi/content-manager/dist/server/mcp/schemas/filters-schema.js
'use strict';

var strapiUtils = require('@strapi/utils');
var sortSchema = require('./sort-schema.js');

/**
 * Maps a scalar Strapi attribute type to the appropriate Zod leaf value schema
 * used inside filter operator objects (e.g. { $eq: <value> }).
 */ const attributeTypeToFilterValue = (attr)=>{
    switch(attr.type){
        case 'integer':
        case 'biginteger':
        case 'decimal':
        case 'float':
            return strapiUtils.z.union([
                strapiUtils.z.number(),
                strapiUtils.z.array(strapiUtils.z.number())
            ]);
        case 'boolean':
            return strapiUtils.z.boolean();
        case 'enumeration':
            {
                const enumAttr = attr;
                if (Array.isArray(enumAttr.enum) && enumAttr.enum.length > 0) {
                    return strapiUtils.z.union([
                        strapiUtils.z.enum(enumAttr.enum),
                        strapiUtils.z.array(strapiUtils.z.enum(enumAttr.enum))
                    ]);
                }
                return strapiUtils.z.union([
                    strapiUtils.z.string(),
                    strapiUtils.z.array(strapiUtils.z.string())
                ]);
            }
        default:
            // string, text, richtext, email, password, uid, date, datetime, time, timestamp
            return strapiUtils.z.union([
                strapiUtils.z.string(),
                strapiUtils.z.array(strapiUtils.z.string()),
                strapiUtils.z.null()
            ]);
    }
};
/** All supported Strapi filter operators (excludes the experimental `$jsonSupersetOf`). */ const FILTER_OPERATORS = [
    '$eq',
    '$eqi',
    '$ne',
    '$nei',
    '$in',
    '$notIn',
    '$lt',
    '$lte',
    '$gt',
    '$gte',
    '$between',
    '$contains',
    '$notContains',
    '$containsi',
    '$notContainsi',
    '$startsWith',
    '$startsWithi',
    '$endsWith',
    '$endsWithi',
    '$null',
    '$notNull'
];
/**
 * Builds a per-content-type recursive filters Zod schema.
 *
 * Shape:
 *   - Logical operators: $and, $or accept an array of filter objects.
 *   - Logical operator: $not accepts a single filter object.
 *   - Field keys (scalar attrs only): accept either a direct value (implicit $eq)
 *     or an operator object { $eq, $contains, $gt, … }.
 *
 * If the model has no scalar attributes, the schema is z.never() (filters not allowed).
 */ const buildFiltersSchema = (attributes, permittedFields)=>{
    const scalarKeys = sortSchema.getScalarAttributeKeys(attributes, permittedFields);
    if (scalarKeys.length === 0) {
        return strapiUtils.z.never();
    }
    // Lazy reference for recursion ($and / $or / $not)
    const filtersSchema = strapiUtils.z.lazy(()=>{
        const fieldShapes = {};
        for (const key of scalarKeys){
            const attr = attributes[key];
            const valueSchema = attributeTypeToFilterValue(attr);
            const operatorObject = strapiUtils.z.object(Object.fromEntries(FILTER_OPERATORS.map((op)=>[
                    op,
                    valueSchema.optional()
                ])));
            // Field accepts either a direct value (implicit $eq) or operator object
            fieldShapes[key] = strapiUtils.z.union([
                valueSchema,
                operatorObject
            ]).optional();
        }
        return strapiUtils.z.object({
            $and: strapiUtils.z.array(filtersSchema).optional(),
            $or: strapiUtils.z.array(filtersSchema).optional(),
            $not: filtersSchema.optional(),
            ...fieldShapes
        }).strict();
    });
    return filtersSchema.optional().describe(`Filter object. Supports logical operators ($and, $or, $not) and field operators ` + `($eq, $ne, $in, $contains, $gt, $lt, etc.). Valid fields: ${scalarKeys.join(', ')}.`);
};

exports.FILTER_OPERATORS = FILTER_OPERATORS;
exports.attributeTypeToFilterValue = attributeTypeToFilterValue;
exports.buildFiltersSchema = buildFiltersSchema;
