// @strapi/review-workflows/dist/admin/index.mjs
import { SealCheck } from '@strapi/icons';
import { FEATURE_ID, PLUGIN_ID } from './constants.mjs';
import { Header } from './routes/content-manager/model/id/components/Header.mjs';
import { Panel } from './routes/content-manager/model/id/components/Panel.mjs';
import { addColumnToTableHook, addFilterToListViewHook } from './utils/cm-hooks.mjs';
import { prefixPluginTranslations } from './utils/translations.mjs';

function __variableDynamicImportRuntime3__(path) {
  switch (path) {
    case './translations/cs.json': return import('./translations/cs.json.mjs');
    case './translations/de.json': return import('./translations/de.json.mjs');
    case './translations/en.json': return import('./translations/en.json.mjs');
    case './translations/es.json': return import('./translations/es.json.mjs');
    case './translations/fr.json': return import('./translations/fr.json.mjs');
    case './translations/nl.json': return import('./translations/nl.json.mjs');
    case './translations/pl.json': return import('./translations/pl.json.mjs');
    case './translations/ru.json': return import('./translations/ru.json.mjs');
    case './translations/sk.json': return import('./translations/sk.json.mjs');
    case './translations/uk.json': return import('./translations/uk.json.mjs');
    case './translations/zh-Hans.json': return import('./translations/zh-Hans.json.mjs');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
const admin = {
    register (app) {
        if (window.strapi.features.isEnabled(FEATURE_ID)) {
            app.registerHook('Admin/CM/pages/ListView/inject-column-in-table', addColumnToTableHook);
            app.registerHook('Admin/CM/pages/ListView/inject-in-filters', addFilterToListViewHook);
            const contentManagerPluginApis = app.getPlugin('content-manager').apis;
            if ('addEditViewSidePanel' in contentManagerPluginApis && typeof contentManagerPluginApis.addEditViewSidePanel === 'function') {
                contentManagerPluginApis.addEditViewSidePanel([
                    Panel
                ]);
            }
            app.addSettingsLink('global', {
                id: PLUGIN_ID,
                to: `review-workflows`,
                intlLabel: {
                    id: `${PLUGIN_ID}.plugin.name`,
                    defaultMessage: 'Review Workflows'
                },
                licenseOnly: true,
                permissions: [],
                Component () {
                    return import('./router.mjs').then((mod)=>({
                            default: mod.Router
                        }));
                }
            });
            // Always put the assigned to me widget last in the list of widgets
            app.widgets.register((widgets)=>[
                    ...widgets,
                    {
                        icon: SealCheck,
                        title: {
                            id: `${PLUGIN_ID}.widget.assigned.title`,
                            defaultMessage: 'Assigned to me'
                        },
                        component: async ()=>{
                            const { AssignedWidget } = await import('./components/Widgets.mjs');
                            return AssignedWidget;
                        },
                        pluginId: PLUGIN_ID,
                        id: 'assigned',
                        permissions: [
                            {
                                action: 'plugin::content-manager.explorer.read'
                            }
                        ]
                    }
                ]);
        } else if (!window.strapi.features.isEnabled(FEATURE_ID) && window.strapi?.flags?.promoteEE) {
            app.addSettingsLink('global', {
                id: PLUGIN_ID,
                to: `purchase-review-workflows`,
                intlLabel: {
                    id: `${PLUGIN_ID}.plugin.name`,
                    defaultMessage: 'Review Workflows'
                },
                licenseOnly: true,
                permissions: [],
                Component () {
                    return import('./routes/purchase-review-workflows.mjs').then((mod)=>({
                            default: mod.PurchaseReviewWorkflows
                        }));
                }
            });
        }
    },
    bootstrap (app) {
        if (window.strapi.features.isEnabled(FEATURE_ID)) {
            app.getPlugin('content-manager').injectComponent('preview', 'actions', {
                name: 'review-workflows-assignee',
                Component: Header
            });
        }
    },
    async registerTrads ({ locales }) {
        const importedTrads = await Promise.all(locales.map((locale)=>{
            return __variableDynamicImportRuntime3__(`./translations/${locale}.json`).then(({ default: data })=>{
                return {
                    data: prefixPluginTranslations(data, PLUGIN_ID),
                    locale
                };
            }).catch(()=>{
                return {
                    data: {},
                    locale
                };
            });
        }));
        return Promise.resolve(importedTrads);
    }
};

export { admin as default };
