// @strapi/admin/dist/admin/admin/src/pages/Settings/pages/AdminTokens/EditView/EditViewPage.mjs
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import * as React from 'react';
import { Flex, Box } from '@strapi/design-system';
import { Formik, Form } from 'formik';
import { useIntl } from 'react-intl';
import { useLocation, useMatch, useNavigate } from 'react-router-dom';
import { Layouts } from '../../../../../components/Layouts/Layout.mjs';
import { Page } from '../../../../../components/PageHelpers.mjs';
import { useTypedSelector } from '../../../../../core/store/hooks.mjs';
import { useAuth } from '../../../../../features/Auth.mjs';
import { useNotification } from '../../../../../features/Notifications.mjs';
import { useTracking } from '../../../../../features/Tracking.mjs';
import { useAPIErrorHandler } from '../../../../../hooks/useAPIErrorHandler.mjs';
import { useRBAC } from '../../../../../hooks/useRBAC.mjs';
import { useGetAdminTokenQuery, useCreateAdminTokenMutation, useUpdateAdminTokenMutation } from '../../../../../services/apiTokens.mjs';
import { isBaseQueryError } from '../../../../../utils/baseQuery.mjs';
import { API_TOKEN_TYPE, apiTokenInformationSchema } from '../../../components/Tokens/constants.mjs';
import { FormApiTokenContainer } from '../../../components/Tokens/FormApiTokenContainer.mjs';
import { FormHead } from '../../../components/Tokens/FormHead.mjs';
import { ApiTokenBox } from '../../../components/Tokens/TokenBox.mjs';
import { AdminPermissions } from './components/AdminPermissions.mjs';

const getOwnerId = (owner)=>{
    if (owner === undefined || owner === null) return null;
    return typeof owner === 'object' ? owner.id : owner;
};
const isCurrentUserTokenOwner = (apiToken, currentUserId)=>{
    if (apiToken === null || apiToken.kind !== 'admin') return true;
    const ownerId = getOwnerId(apiToken.adminUserOwner);
    if (ownerId === null) return true;
    return currentUserId !== undefined && ownerId === currentUserId;
};
const EditView = ()=>{
    const { formatMessage } = useIntl();
    const { toggleNotification } = useNotification();
    const { state: locationState } = useLocation();
    const permissions = useTypedSelector((state)=>state.admin_app.permissions);
    const [apiToken, setApiToken] = React.useState(locationState?.apiToken?.accessKey !== undefined && locationState?.apiToken?.accessKey !== '' ? {
        ...locationState.apiToken
    } : null);
    const [showToken, setShowToken] = React.useState(locationState?.apiToken?.accessKey !== undefined && locationState?.apiToken?.accessKey !== '');
    const hideTimerRef = React.useRef(null);
    const adminPermissionsRef = React.useRef(null);
    const { trackUsage } = useTracking();
    const currentUser = useAuth('EditView', (state)=>state.user);
    const { allowedActions: { canCreate, canUpdate, canRegenerate } } = useRBAC(permissions.settings?.['admin-tokens']);
    const match = useMatch('/settings/admin-tokens/:id');
    const id = match?.params?.id;
    const isCreating = id === 'create';
    const { _unstableFormatAPIError: formatAPIError, _unstableFormatValidationErrors: formatValidationErrors } = useAPIErrorHandler();
    const navigate = useNavigate();
    React.useEffect(()=>{
        trackUsage(isCreating ? 'didAddTokenFromList' : 'didEditTokenFromList', {
            tokenType: API_TOKEN_TYPE
        });
    }, [
        isCreating,
        trackUsage
    ]);
    const { data, error, isLoading } = useGetAdminTokenQuery(id, {
        skip: !id || isCreating,
        refetchOnMountOrArgChange: true
    });
    React.useEffect(()=>{
        if (error) {
            toggleNotification({
                type: 'danger',
                message: formatAPIError(error)
            });
        }
    }, [
        error,
        formatAPIError,
        toggleNotification
    ]);
    React.useEffect(()=>{
        if (data) {
            setApiToken(data);
        }
    }, [
        data
    ]);
    React.useEffect(()=>{
        if (showToken) {
            hideTimerRef.current = setTimeout(()=>{
                setShowToken(false);
            }, 30000); // 30 seconds
            return ()=>{
                if (hideTimerRef.current) {
                    clearTimeout(hideTimerRef.current);
                    hideTimerRef.current = null;
                }
            };
        }
    }, [
        showToken
    ]);
    const [createToken] = useCreateAdminTokenMutation();
    const [updateToken] = useUpdateAdminTokenMutation();
    const buildLifespan = (raw)=>raw && raw !== '0' ? parseInt(raw.toString(), 10) : null;
    const handleSubmit = async (body, formik)=>{
        trackUsage(isCreating ? 'willCreateToken' : 'willEditToken', {
            tokenType: API_TOKEN_TYPE
        });
        const adminPermissionsToSend = adminPermissionsRef.current?.getPermissions().permissionsToSend ?? [];
        try {
            if (isCreating) {
                const res = await createToken({
                    name: body.name,
                    description: body.description,
                    lifespan: buildLifespan(body.lifespan),
                    adminPermissions: adminPermissionsToSend
                });
                if ('error' in res) {
                    if (isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                        formik.setErrors(formatValidationErrors(res.error));
                    } else {
                        toggleNotification({
                            type: 'danger',
                            message: formatAPIError(res.error)
                        });
                    }
                    return;
                }
                toggleNotification({
                    type: 'success',
                    message: formatMessage({
                        id: 'notification.success.adminTokenCreated',
                        defaultMessage: 'Admin Token successfully created'
                    })
                });
                trackUsage('didCreateToken', {
                    kind: 'admin',
                    tokenType: API_TOKEN_TYPE
                });
                adminPermissionsRef.current?.setFormAfterSubmit();
                navigate(`../admin-tokens/${res.data.id.toString()}`, {
                    state: {
                        apiToken: res.data
                    },
                    replace: true
                });
            } else {
                const res = await updateToken({
                    id: id,
                    name: body.name,
                    description: body.description,
                    adminPermissions: adminPermissionsToSend
                });
                if ('error' in res) {
                    if (isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                        formik.setErrors(formatValidationErrors(res.error));
                    } else {
                        toggleNotification({
                            type: 'danger',
                            message: formatAPIError(res.error)
                        });
                    }
                    return;
                }
                toggleNotification({
                    type: 'success',
                    message: formatMessage({
                        id: 'notification.success.adminTokenEdited',
                        defaultMessage: 'Admin Token successfully edited'
                    })
                });
                trackUsage('didEditToken', {
                    kind: 'admin',
                    tokenType: API_TOKEN_TYPE
                });
                adminPermissionsRef.current?.setFormAfterSubmit();
            }
        } catch  {
            toggleNotification({
                type: 'danger',
                message: formatMessage({
                    id: 'notification.error',
                    defaultMessage: 'Something went wrong'
                })
            });
        }
    };
    const toggleToken = ()=>{
        setShowToken((prev)=>!prev);
        if (hideTimerRef.current) {
            clearTimeout(hideTimerRef.current);
            hideTimerRef.current = null;
        }
    };
    const canEditInputs = canUpdate && !isCreating || canCreate && isCreating;
    const canShowToken = apiToken?.accessKey !== undefined && apiToken.accessKey !== '';
    const canRegenerateToken = canRegenerate && isCurrentUserTokenOwner(apiToken, currentUser?.id);
    const initialAdminPermissions = apiToken !== null && apiToken.kind === 'admin' ? apiToken.adminPermissions ?? [] : [];
    const ownerUserId = apiToken !== null && apiToken.kind === 'admin' ? getOwnerId(apiToken.adminUserOwner) : null;
    if (isLoading || !isCreating && apiToken === null) {
        return /*#__PURE__*/ jsx(Page.Loading, {});
    }
    return /*#__PURE__*/ jsxs(Page.Main, {
        children: [
            /*#__PURE__*/ jsx(Page.Title, {
                children: formatMessage({
                    id: 'Settings.PageTitle',
                    defaultMessage: 'Settings - {name}'
                }, {
                    name: 'Admin Tokens'
                })
            }),
            /*#__PURE__*/ jsx(Formik, {
                validationSchema: apiTokenInformationSchema,
                validateOnChange: false,
                initialValues: {
                    name: apiToken?.name || '',
                    description: apiToken?.description || '',
                    type: undefined,
                    lifespan: apiToken?.lifespan
                },
                enableReinitialize: true,
                onSubmit: (body, actions)=>handleSubmit(body, actions),
                children: ({ errors, handleChange, isSubmitting, values })=>/*#__PURE__*/ jsxs(Form, {
                        children: [
                            /*#__PURE__*/ jsx(FormHead, {
                                title: isCreating ? {
                                    id: 'Settings.adminTokens.createPage.title',
                                    defaultMessage: 'Create Admin Token'
                                } : {
                                    id: 'Settings.adminTokens.editPage.title',
                                    defaultMessage: 'Edit Admin Token'
                                },
                                token: apiToken,
                                setToken: setApiToken,
                                toggleToken: toggleToken,
                                showToken: showToken,
                                canEditInputs: canEditInputs,
                                canRegenerate: canRegenerateToken,
                                canShowToken: canShowToken,
                                isSubmitting: isSubmitting,
                                regenerateUrl: "/admin/admin-tokens/"
                            }),
                            /*#__PURE__*/ jsx(Layouts.Content, {
                                children: /*#__PURE__*/ jsxs(Flex, {
                                    direction: "column",
                                    alignItems: "stretch",
                                    gap: 6,
                                    children: [
                                        apiToken?.accessKey !== undefined && apiToken.accessKey !== '' && showToken === true && /*#__PURE__*/ jsx(Fragment, {
                                            children: /*#__PURE__*/ jsx(ApiTokenBox, {
                                                token: apiToken.accessKey,
                                                tokenType: API_TOKEN_TYPE
                                            })
                                        }),
                                        /*#__PURE__*/ jsx(FormApiTokenContainer, {
                                            errors: errors,
                                            onChange: handleChange,
                                            canEditInputs: canEditInputs,
                                            isCreating: isCreating,
                                            values: values,
                                            apiToken: apiToken,
                                            kind: "admin",
                                            onDispatch: ()=>{},
                                            setHasChangedPermissions: ()=>{}
                                        }),
                                        /*#__PURE__*/ jsx(Box, {
                                            shadow: "filterShadow",
                                            hasRadius: true,
                                            children: /*#__PURE__*/ jsx(AdminPermissions, {
                                                ref: adminPermissionsRef,
                                                disabled: !canEditInputs,
                                                initialAdminPermissions: initialAdminPermissions,
                                                tokenId: !isCreating && id ? id : undefined,
                                                ownerUserId: ownerUserId
                                            })
                                        })
                                    ]
                                })
                            })
                        ]
                    })
            })
        ]
    });
};
const ProtectedEditView = ()=>{
    const permissions = useTypedSelector((state)=>state.admin_app.permissions.settings?.['admin-tokens']?.read);
    return /*#__PURE__*/ jsx(Page.Protect, {
        permissions: permissions,
        children: /*#__PURE__*/ jsx(EditView, {})
    });
};

export { EditView, ProtectedEditView };
