// @strapi/admin/dist/admin/admin/src/pages/Settings/pages/AdminTokens/EditView/EditViewPage.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var formik = require('formik');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var Layout = require('../../../../../components/Layouts/Layout.js');
var PageHelpers = require('../../../../../components/PageHelpers.js');
var hooks = require('../../../../../core/store/hooks.js');
var Auth = require('../../../../../features/Auth.js');
var Notifications = require('../../../../../features/Notifications.js');
var Tracking = require('../../../../../features/Tracking.js');
var useAPIErrorHandler = require('../../../../../hooks/useAPIErrorHandler.js');
var useRBAC = require('../../../../../hooks/useRBAC.js');
var apiTokens = require('../../../../../services/apiTokens.js');
var baseQuery = require('../../../../../utils/baseQuery.js');
var constants = require('../../../components/Tokens/constants.js');
var FormApiTokenContainer = require('../../../components/Tokens/FormApiTokenContainer.js');
var FormHead = require('../../../components/Tokens/FormHead.js');
var TokenBox = require('../../../components/Tokens/TokenBox.js');
var AdminPermissions = require('./components/AdminPermissions.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const getOwnerId = (owner)=>{
    if (owner === undefined || owner === null) return null;
    return typeof owner === 'object' ? owner.id : owner;
};
const isCurrentUserTokenOwner = (apiToken, currentUserId)=>{
    if (apiToken === null || apiToken.kind !== 'admin') return true;
    const ownerId = getOwnerId(apiToken.adminUserOwner);
    if (ownerId === null) return true;
    return currentUserId !== undefined && ownerId === currentUserId;
};
const EditView = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const { toggleNotification } = Notifications.useNotification();
    const { state: locationState } = reactRouterDom.useLocation();
    const permissions = hooks.useTypedSelector((state)=>state.admin_app.permissions);
    const [apiToken, setApiToken] = React__namespace.useState(locationState?.apiToken?.accessKey !== undefined && locationState?.apiToken?.accessKey !== '' ? {
        ...locationState.apiToken
    } : null);
    const [showToken, setShowToken] = React__namespace.useState(locationState?.apiToken?.accessKey !== undefined && locationState?.apiToken?.accessKey !== '');
    const hideTimerRef = React__namespace.useRef(null);
    const adminPermissionsRef = React__namespace.useRef(null);
    const { trackUsage } = Tracking.useTracking();
    const currentUser = Auth.useAuth('EditView', (state)=>state.user);
    const { allowedActions: { canCreate, canUpdate, canRegenerate } } = useRBAC.useRBAC(permissions.settings?.['admin-tokens']);
    const match = reactRouterDom.useMatch('/settings/admin-tokens/:id');
    const id = match?.params?.id;
    const isCreating = id === 'create';
    const { _unstableFormatAPIError: formatAPIError, _unstableFormatValidationErrors: formatValidationErrors } = useAPIErrorHandler.useAPIErrorHandler();
    const navigate = reactRouterDom.useNavigate();
    React__namespace.useEffect(()=>{
        trackUsage(isCreating ? 'didAddTokenFromList' : 'didEditTokenFromList', {
            tokenType: constants.API_TOKEN_TYPE
        });
    }, [
        isCreating,
        trackUsage
    ]);
    const { data, error, isLoading } = apiTokens.useGetAdminTokenQuery(id, {
        skip: !id || isCreating,
        refetchOnMountOrArgChange: true
    });
    React__namespace.useEffect(()=>{
        if (error) {
            toggleNotification({
                type: 'danger',
                message: formatAPIError(error)
            });
        }
    }, [
        error,
        formatAPIError,
        toggleNotification
    ]);
    React__namespace.useEffect(()=>{
        if (data) {
            setApiToken(data);
        }
    }, [
        data
    ]);
    React__namespace.useEffect(()=>{
        if (showToken) {
            hideTimerRef.current = setTimeout(()=>{
                setShowToken(false);
            }, 30000); // 30 seconds
            return ()=>{
                if (hideTimerRef.current) {
                    clearTimeout(hideTimerRef.current);
                    hideTimerRef.current = null;
                }
            };
        }
    }, [
        showToken
    ]);
    const [createToken] = apiTokens.useCreateAdminTokenMutation();
    const [updateToken] = apiTokens.useUpdateAdminTokenMutation();
    const buildLifespan = (raw)=>raw && raw !== '0' ? parseInt(raw.toString(), 10) : null;
    const handleSubmit = async (body, formik)=>{
        trackUsage(isCreating ? 'willCreateToken' : 'willEditToken', {
            tokenType: constants.API_TOKEN_TYPE
        });
        const adminPermissionsToSend = adminPermissionsRef.current?.getPermissions().permissionsToSend ?? [];
        try {
            if (isCreating) {
                const res = await createToken({
                    name: body.name,
                    description: body.description,
                    lifespan: buildLifespan(body.lifespan),
                    adminPermissions: adminPermissionsToSend
                });
                if ('error' in res) {
                    if (baseQuery.isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                        formik.setErrors(formatValidationErrors(res.error));
                    } else {
                        toggleNotification({
                            type: 'danger',
                            message: formatAPIError(res.error)
                        });
                    }
                    return;
                }
                toggleNotification({
                    type: 'success',
                    message: formatMessage({
                        id: 'notification.success.adminTokenCreated',
                        defaultMessage: 'Admin Token successfully created'
                    })
                });
                trackUsage('didCreateToken', {
                    kind: 'admin',
                    tokenType: constants.API_TOKEN_TYPE
                });
                adminPermissionsRef.current?.setFormAfterSubmit();
                navigate(`../admin-tokens/${res.data.id.toString()}`, {
                    state: {
                        apiToken: res.data
                    },
                    replace: true
                });
            } else {
                const res = await updateToken({
                    id: id,
                    name: body.name,
                    description: body.description,
                    adminPermissions: adminPermissionsToSend
                });
                if ('error' in res) {
                    if (baseQuery.isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                        formik.setErrors(formatValidationErrors(res.error));
                    } else {
                        toggleNotification({
                            type: 'danger',
                            message: formatAPIError(res.error)
                        });
                    }
                    return;
                }
                toggleNotification({
                    type: 'success',
                    message: formatMessage({
                        id: 'notification.success.adminTokenEdited',
                        defaultMessage: 'Admin Token successfully edited'
                    })
                });
                trackUsage('didEditToken', {
                    kind: 'admin',
                    tokenType: constants.API_TOKEN_TYPE
                });
                adminPermissionsRef.current?.setFormAfterSubmit();
            }
        } catch  {
            toggleNotification({
                type: 'danger',
                message: formatMessage({
                    id: 'notification.error',
                    defaultMessage: 'Something went wrong'
                })
            });
        }
    };
    const toggleToken = ()=>{
        setShowToken((prev)=>!prev);
        if (hideTimerRef.current) {
            clearTimeout(hideTimerRef.current);
            hideTimerRef.current = null;
        }
    };
    const canEditInputs = canUpdate && !isCreating || canCreate && isCreating;
    const canShowToken = apiToken?.accessKey !== undefined && apiToken.accessKey !== '';
    const canRegenerateToken = canRegenerate && isCurrentUserTokenOwner(apiToken, currentUser?.id);
    const initialAdminPermissions = apiToken !== null && apiToken.kind === 'admin' ? apiToken.adminPermissions ?? [] : [];
    const ownerUserId = apiToken !== null && apiToken.kind === 'admin' ? getOwnerId(apiToken.adminUserOwner) : null;
    if (isLoading || !isCreating && apiToken === null) {
        return /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Loading, {});
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(PageHelpers.Page.Main, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Title, {
                children: formatMessage({
                    id: 'Settings.PageTitle',
                    defaultMessage: 'Settings - {name}'
                }, {
                    name: 'Admin Tokens'
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(formik.Formik, {
                validationSchema: constants.apiTokenInformationSchema,
                validateOnChange: false,
                initialValues: {
                    name: apiToken?.name || '',
                    description: apiToken?.description || '',
                    type: undefined,
                    lifespan: apiToken?.lifespan
                },
                enableReinitialize: true,
                onSubmit: (body, actions)=>handleSubmit(body, actions),
                children: ({ errors, handleChange, isSubmitting, values })=>/*#__PURE__*/ jsxRuntime.jsxs(formik.Form, {
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(FormHead.FormHead, {
                                title: isCreating ? {
                                    id: 'Settings.adminTokens.createPage.title',
                                    defaultMessage: 'Create Admin Token'
                                } : {
                                    id: 'Settings.adminTokens.editPage.title',
                                    defaultMessage: 'Edit Admin Token'
                                },
                                token: apiToken,
                                setToken: setApiToken,
                                toggleToken: toggleToken,
                                showToken: showToken,
                                canEditInputs: canEditInputs,
                                canRegenerate: canRegenerateToken,
                                canShowToken: canShowToken,
                                isSubmitting: isSubmitting,
                                regenerateUrl: "/admin/admin-tokens/"
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(Layout.Layouts.Content, {
                                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                    direction: "column",
                                    alignItems: "stretch",
                                    gap: 6,
                                    children: [
                                        apiToken?.accessKey !== undefined && apiToken.accessKey !== '' && showToken === true && /*#__PURE__*/ jsxRuntime.jsx(jsxRuntime.Fragment, {
                                            children: /*#__PURE__*/ jsxRuntime.jsx(TokenBox.ApiTokenBox, {
                                                token: apiToken.accessKey,
                                                tokenType: constants.API_TOKEN_TYPE
                                            })
                                        }),
                                        /*#__PURE__*/ jsxRuntime.jsx(FormApiTokenContainer.FormApiTokenContainer, {
                                            errors: errors,
                                            onChange: handleChange,
                                            canEditInputs: canEditInputs,
                                            isCreating: isCreating,
                                            values: values,
                                            apiToken: apiToken,
                                            kind: "admin",
                                            onDispatch: ()=>{},
                                            setHasChangedPermissions: ()=>{}
                                        }),
                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                            shadow: "filterShadow",
                                            hasRadius: true,
                                            children: /*#__PURE__*/ jsxRuntime.jsx(AdminPermissions.AdminPermissions, {
                                                ref: adminPermissionsRef,
                                                disabled: !canEditInputs,
                                                initialAdminPermissions: initialAdminPermissions,
                                                tokenId: !isCreating && id ? id : undefined,
                                                ownerUserId: ownerUserId
                                            })
                                        })
                                    ]
                                })
                            })
                        ]
                    })
            })
        ]
    });
};
const ProtectedEditView = ()=>{
    const permissions = hooks.useTypedSelector((state)=>state.admin_app.permissions.settings?.['admin-tokens']?.read);
    return /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Protect, {
        permissions: permissions,
        children: /*#__PURE__*/ jsxRuntime.jsx(EditView, {})
    });
};

exports.EditView = EditView;
exports.ProtectedEditView = ProtectedEditView;
