// @strapi/review-workflows/dist/admin/routes/content-manager/model/id/components/AssigneeSelect.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { useAPIErrorHandler, useNotification, useRBAC, useQueryParams, useDebounce, useAdminUsers } from '@strapi/admin/strapi-admin';
import { unstable_useDocument } from '@strapi/content-manager/strapi-admin';
import { Field, VisuallyHidden, Combobox, ComboboxOption } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { useParams } from 'react-router-dom';
import { useTypedSelector } from '../../../../../modules/hooks.mjs';
import { useUpdateAssigneeMutation } from '../../../../../services/content-manager.mjs';
import { buildValidParams } from '../../../../../utils/api.mjs';
import { getDisplayName } from '../../../../../utils/users.mjs';
import { ASSIGNEE_ATTRIBUTE_NAME } from './constants.mjs';

const PAGE_SIZE = 10;
const contains = (value)=>({
        $containsi: value
    });
const AssigneeSelect = ({ isCompact })=>{
    const { collectionType = '', id, slug: model = '' } = useParams();
    const permissions = useTypedSelector((state)=>state.admin_app.permissions);
    const { formatMessage } = useIntl();
    const { _unstableFormatAPIError: formatAPIError } = useAPIErrorHandler();
    const { toggleNotification } = useNotification();
    const { allowedActions: { canRead }, isLoading: isLoadingPermissions } = useRBAC(permissions.settings?.users);
    const [{ query }] = useQueryParams();
    const params = React.useMemo(()=>buildValidParams(query), [
        query
    ]);
    const [pageSize, setPageSize] = React.useState(PAGE_SIZE);
    const [search, setSearch] = React.useState('');
    const debouncedSearch = useDebounce(search, 300);
    const searchFilters = React.useMemo(()=>{
        const value = debouncedSearch.trim();
        if (!value) {
            return undefined;
        }
        const [firstTerm, ...restTerms] = value.split(/\s+/);
        const rest = restTerms.join(' ');
        const filters = {
            $or: [
                {
                    firstname: contains(value)
                },
                {
                    lastname: contains(value)
                },
                {
                    username: contains(value)
                },
                {
                    email: contains(value)
                }
            ]
        };
        if (rest) {
            filters.$or = [
                ...filters.$or ?? [],
                {
                    $and: [
                        {
                            firstname: contains(firstTerm)
                        },
                        {
                            lastname: contains(rest)
                        }
                    ]
                },
                {
                    $and: [
                        {
                            firstname: contains(rest)
                        },
                        {
                            lastname: contains(firstTerm)
                        }
                    ]
                }
            ];
        }
        return filters;
    }, [
        debouncedSearch
    ]);
    const { data, isLoading: isLoadingUsers, isError } = useAdminUsers({
        pageSize,
        filters: searchFilters
    }, {
        skip: isLoadingPermissions || !canRead
    });
    const { document } = unstable_useDocument({
        collectionType,
        model,
        documentId: id,
        params
    }, {
        skip: !id && collectionType !== 'single-types'
    });
    const users = React.useMemo(()=>data?.users ?? [], [
        data?.users
    ]);
    const { pageCount = 1, page = 1 } = data?.pagination ?? {};
    const currentAssignee = document ? document[ASSIGNEE_ATTRIBUTE_NAME] : null;
    // Keep the currently assigned user in the options even when they fall outside
    // the loaded page or the active search — otherwise the Combobox loses its value.
    const options = React.useMemo(()=>{
        if (!currentAssignee) return users;
        return users.some((u)=>u.id === currentAssignee.id) ? users : [
            currentAssignee,
            ...users
        ];
    }, [
        users,
        currentAssignee
    ]);
    const handleOpenChange = (isOpen)=>{
        if (!isOpen) {
            setPageSize(PAGE_SIZE);
            setSearch('');
        }
    };
    const handleLoadMore = ()=>{
        setPageSize(pageSize + PAGE_SIZE);
    };
    const handleInputChange = (e)=>{
        setSearch(e.currentTarget.value);
        setPageSize(PAGE_SIZE);
    };
    const [updateAssignee, { error, isLoading: isMutating }] = useUpdateAssigneeMutation();
    if (!collectionType || !model || !document?.documentId) {
        return null;
    }
    const handleChange = async (assigneeId)=>{
        // a simple way to avoid erroneous updates
        if (currentAssignee?.id === assigneeId) {
            return;
        }
        const res = await updateAssignee({
            slug: collectionType,
            model,
            id: document.documentId,
            params,
            data: {
                id: assigneeId ? parseInt(assigneeId, 10) : null
            }
        });
        if ('data' in res) {
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: 'review-workflows.assignee.notification.saved',
                    defaultMessage: 'Assignee updated'
                })
            });
        }
        if (isCompact && 'error' in res) {
            toggleNotification({
                type: 'danger',
                message: formatAPIError(res.error)
            });
        }
    };
    const isDisabled = !isLoadingPermissions && !isLoadingUsers && users.length === 0 || !document.documentId;
    const isLoading = isLoadingUsers || isLoadingPermissions || isMutating;
    const hasMoreItems = page < pageCount;
    const assigneeLabel = formatMessage({
        id: 'review-workflows.assignee.label',
        defaultMessage: 'Assignee'
    });
    const assigneeClearLabel = formatMessage({
        id: 'review-workflows.assignee.clear',
        defaultMessage: 'Clear assignee'
    });
    const assigneePlaceholder = formatMessage({
        id: 'review-workflows.assignee.placeholder',
        defaultMessage: 'Select…'
    });
    if (isCompact) {
        return /*#__PURE__*/ jsxs(Field.Root, {
            name: ASSIGNEE_ATTRIBUTE_NAME,
            id: ASSIGNEE_ATTRIBUTE_NAME,
            children: [
                /*#__PURE__*/ jsx(VisuallyHidden, {
                    children: /*#__PURE__*/ jsx(Field.Label, {
                        children: assigneeLabel
                    })
                }),
                /*#__PURE__*/ jsx(Combobox, {
                    clearLabel: assigneeClearLabel,
                    disabled: isDisabled,
                    value: currentAssignee ? currentAssignee.id.toString() : null,
                    onChange: handleChange,
                    onClear: ()=>handleChange(null),
                    onOpenChange: handleOpenChange,
                    onLoadMore: handleLoadMore,
                    hasMoreItems: hasMoreItems,
                    onInputChange: handleInputChange,
                    placeholder: assigneePlaceholder,
                    loading: isLoading || isLoadingPermissions || isMutating,
                    size: "S",
                    children: options.map((user)=>{
                        return /*#__PURE__*/ jsx(ComboboxOption, {
                            value: user.id.toString(),
                            textValue: getDisplayName(user),
                            children: getDisplayName(user)
                        }, user.id);
                    })
                })
            ]
        });
    }
    return /*#__PURE__*/ jsxs(Field.Root, {
        name: ASSIGNEE_ATTRIBUTE_NAME,
        id: ASSIGNEE_ATTRIBUTE_NAME,
        error: (isError && canRead && formatMessage({
            id: 'review-workflows.assignee.error',
            defaultMessage: 'An error occurred while fetching users'
        }) || error && formatAPIError(error)) ?? undefined,
        children: [
            /*#__PURE__*/ jsx(Field.Label, {
                children: assigneeLabel
            }),
            /*#__PURE__*/ jsx(Combobox, {
                clearLabel: assigneeClearLabel,
                disabled: !isLoadingPermissions && !isLoading && users.length === 0 || !document.documentId,
                value: currentAssignee ? currentAssignee.id.toString() : null,
                onChange: handleChange,
                onClear: ()=>handleChange(null),
                onOpenChange: handleOpenChange,
                onLoadMore: handleLoadMore,
                hasMoreItems: hasMoreItems,
                onInputChange: handleInputChange,
                placeholder: assigneePlaceholder,
                loading: isLoading || isLoadingPermissions || isMutating,
                children: options.map((user)=>{
                    return /*#__PURE__*/ jsx(ComboboxOption, {
                        value: user.id.toString(),
                        textValue: getDisplayName(user),
                        children: getDisplayName(user)
                    }, user.id);
                })
            }),
            /*#__PURE__*/ jsx(Field.Error, {})
        ]
    });
};

export { AssigneeSelect };
