// @strapi/core/dist/services/mcp/handlers/handlePost.js
'use strict';

var streamableHttp_js = require('@modelcontextprotocol/sdk/server/streamableHttp.js');
var metrics = require('../metrics/metrics.js');
var sendJsonRpcError = require('../utils/sendJsonRpcError.js');
var withTimeout = require('../utils/withTimeout.js');

// eslint-disable-next-line import/extensions
const createPostHandler = (deps)=>{
    const { strapi, authenticationStrategy, config, createServerWithRegistries, capabilityDefinitions } = deps;
    return async (ctx)=>{
        // Opt out of Koa's response phase — the MCP SDK writes directly to ctx.res
        // (via res.writeHead / res.end / SSE streaming). Without this, Koa's respond()
        // would also try to write ctx.body to the socket after the handler returns.
        ctx.respond = false;
        const req = ctx.req;
        const res = ctx.res;
        let hadAuthenticatedMcpRequest = false;
        try {
            const authResult = await authenticationStrategy.authenticate(ctx);
            if (authResult.authenticated === false) {
                metrics.sendDidNotAuthenticateMcpRequest(strapi, authResult.reason);
                sendJsonRpcError.sendJsonRpcError(res, 'AUTHENTICATION_REQUIRED');
                return;
            }
            hadAuthenticatedMcpRequest = true;
            metrics.sendDidUseMcpServer(strapi);
            const { mcpServer } = createServerWithRegistries({
                strapi,
                definitions: capabilityDefinitions,
                isDevMode: config.isDevMode(),
                ability: authResult.ability,
                user: authResult.user
            });
            const transport = new streamableHttp_js.StreamableHTTPServerTransport({
                sessionIdGenerator: undefined
            });
            try {
                await withTimeout.withTimeout(mcpServer.connect(transport), config.connectTimeoutMs, 'mcpServer.connect');
                const requestBody = ctx.request.body ?? null;
                await withTimeout.withTimeout(transport.handleRequest(req, res, requestBody), config.requestTimeoutMs, 'transport.handleRequest');
            } finally{
                await mcpServer.close();
            }
        } catch (error) {
            strapi.log.error('[MCP] Error handling POST request', {
                error: error instanceof Error ? error.message : 'Unknown error',
                stack: error instanceof Error ? error.stack : undefined
            });
            sendJsonRpcError.sendJsonRpcError(res, 'INTERNAL_ERROR');
            if (hadAuthenticatedMcpRequest) {
                metrics.sendDidNotHandleMcpRequest(strapi, metrics.classifyMcpRequestFailure(error));
            }
        }
    };
};

exports.createPostHandler = createPostHandler;
