// @strapi/admin/dist/admin/admin/src/components/UpsellBanner.mjs
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { useEffect } from 'react';
import { useLicenseLimits } from '@strapi/admin/strapi-admin/ee';
import { IconButton, Box, Flex, Typography, LinkButton } from '@strapi/design-system';
import { ArrowsOut, Cross } from '@strapi/icons';
import { subDays, isAfter, isValid } from 'date-fns';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';
import { useGetLicenseTrialTimeLeftQuery } from '../services/admin.mjs';
import { RESPONSIVE_DEFAULT_SPACING } from '../constants/theme.mjs';
import { useScopedPersistentState } from '../hooks/usePersistentState.mjs';

const BannerBackground = styled(Flex)`
  background: linear-gradient(
    90deg,
    ${({ theme })=>theme.colors.primary600} 0%,
    ${({ theme })=>theme.colors.alternative600} 121.48%
  );
  position: relative;
`;
const FixedButtonWrapper = styled(Box)`
  position: fixed;
  display: flex;
  flex-direction: column;
  z-index: 11;
  align-items: flex-end;
  top: 9px;
  right: 16px;
`;
const Banner = ({ isTrialEndedRecently, onDismiss })=>{
    const { formatMessage } = useIntl();
    return /*#__PURE__*/ jsxs(Fragment, {
        children: [
            /*#__PURE__*/ jsx(BannerBackground, {
                width: "100%",
                justifyContent: "center",
                children: /*#__PURE__*/ jsxs(Flex, {
                    justifyContent: "center",
                    alignItems: "center",
                    width: "100%",
                    paddingTop: 2,
                    paddingBottom: 2,
                    paddingLeft: RESPONSIVE_DEFAULT_SPACING,
                    paddingRight: RESPONSIVE_DEFAULT_SPACING,
                    gap: 2,
                    children: [
                        /*#__PURE__*/ jsxs(Box, {
                            children: [
                                /*#__PURE__*/ jsx(Typography, {
                                    variant: "delta",
                                    fontWeight: "bold",
                                    textColor: "neutral0",
                                    textAlign: "center",
                                    fontSize: 2,
                                    children: formatMessage(isTrialEndedRecently ? {
                                        id: 'app.components.UpsellBanner.intro.ended',
                                        defaultMessage: 'Your trial has ended: '
                                    } : {
                                        id: 'app.components.UpsellBanner.intro',
                                        defaultMessage: 'Access to Growth plan features: '
                                    })
                                }),
                                /*#__PURE__*/ jsx(Typography, {
                                    variant: "delta",
                                    textColor: "neutral0",
                                    textAlign: "center",
                                    paddingRight: 4,
                                    fontSize: 2,
                                    children: formatMessage(isTrialEndedRecently ? {
                                        id: 'app.components.UpsellBanner.text.ended',
                                        defaultMessage: 'Keep access to Growth features by upgrading now.'
                                    } : {
                                        id: 'app.components.UpsellBanner.text',
                                        defaultMessage: 'As part of your trial, you can explore premium tools such as Content History, Releases, and Single Sign-On (SSO).'
                                    })
                                })
                            ]
                        }),
                        /*#__PURE__*/ jsx(Box, {
                            children: /*#__PURE__*/ jsx(LinkButton, {
                                width: "max-content",
                                variant: "tertiary",
                                href: "https://strapi.chargebeeportal.com",
                                target: "_blank",
                                children: formatMessage(isTrialEndedRecently ? {
                                    id: 'app.components.UpsellBanner.button.ended',
                                    defaultMessage: 'Keep Growth plan'
                                } : {
                                    id: 'app.components.UpsellBanner.button',
                                    defaultMessage: 'Upgrade now'
                                })
                            })
                        })
                    ]
                })
            }),
            /*#__PURE__*/ jsx(FixedButtonWrapper, {
                children: /*#__PURE__*/ jsx(IconButton, {
                    withTooltip: false,
                    label: formatMessage({
                        id: 'app.components.UpsellBanner.close',
                        defaultMessage: 'Close'
                    }),
                    onClick: onDismiss,
                    children: /*#__PURE__*/ jsx(Cross, {})
                })
            })
        ]
    });
};
const UpsellBanner = ()=>{
    const { license } = useLicenseLimits();
    const { formatMessage } = useIntl();
    const [cachedTrialEndsAt, setCachedTrialEndsAt] = useScopedPersistentState('STRAPI_FREE_TRIAL_ENDS_AT', undefined);
    const [dismissedFor, setDismissedFor] = useScopedPersistentState('STRAPI_UPSELL_BANNER_DISMISSED_FOR', undefined);
    const sevenDaysAgo = subDays(new Date(), 7);
    const timeLeftData = useGetLicenseTrialTimeLeftQuery(undefined, {
        skip: !license?.isTrial
    });
    useEffect(()=>{
        if (timeLeftData.data?.trialEndsAt) {
            setCachedTrialEndsAt(timeLeftData.data.trialEndsAt);
        }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [
        timeLeftData.data?.trialEndsAt
    ]);
    // When the license is not a trial + not EE, and the cached trial end date is found in the localstorage, that means the trial has ended
    // We show the banner to encourage the user to upgrade (for 7 days after the trial ends)
    const isTrialEndedRecently = Boolean(!license?.isTrial && !window.strapi.isEE && cachedTrialEndsAt && isAfter(new Date(cachedTrialEndsAt), sevenDaysAgo));
    const trialEndsAt = timeLeftData.data?.trialEndsAt ?? cachedTrialEndsAt;
    const toCanonicalISO = (v)=>{
        if (!v) return undefined;
        const date = new Date(v);
        return isValid(date) ? date.toISOString() : undefined;
    };
    const isDismissed = Boolean(trialEndsAt && toCanonicalISO(dismissedFor) === toCanonicalISO(trialEndsAt));
    const handleDismiss = ()=>setDismissedFor(toCanonicalISO(trialEndsAt));
    const handleReopen = ()=>setDismissedFor(undefined);
    if (!(timeLeftData.data?.trialEndsAt || isTrialEndedRecently)) {
        return null;
    }
    if (isDismissed) {
        return /*#__PURE__*/ jsx(FixedButtonWrapper, {
            children: /*#__PURE__*/ jsx(IconButton, {
                withTooltip: false,
                label: formatMessage({
                    id: 'app.components.UpsellBanner.reopen',
                    defaultMessage: 'Reopen banner'
                }),
                onClick: handleReopen,
                children: /*#__PURE__*/ jsx(ArrowsOut, {})
            })
        });
    }
    return /*#__PURE__*/ jsx(Banner, {
        isTrialEndedRecently: isTrialEndedRecently,
        onDismiss: handleDismiss
    });
};

export { UpsellBanner };
