// @strapi/core/dist/services/content-source-maps.js
'use strict';

var stega = require('@vercel/stega');
var strapiUtils = require('@strapi/utils');

const ENCODABLE_TYPES = [
    'string',
    'text',
    'richtext',
    'biginteger',
    'date',
    'time',
    'datetime',
    'timestamp',
    'boolean',
    'enumeration',
    'json',
    'media',
    'email',
    'password'
];
// TODO: use a centralized store for these fields that would be shared with the CM and CTB
const EXCLUDED_FIELDS = [
    'id',
    'documentId',
    'locale',
    'localizations',
    'created_by',
    'updated_by',
    'created_at',
    'updated_at',
    'publishedAt'
];
const isObject = (value)=>{
    return typeof value === 'object' && value !== null;
};
const createContentSourceMapsService = (strapi)=>{
    return {
        encodeField (text, { kind, model, documentId, type, path, locale }) {
            /**
       * Combine all metadata into into a one string so we only have to deal with one data-atribute
       * on the frontend. Make it human readable because that data-attribute may be set manually by
       * users for fields that don't support sourcemap encoding.
       */ const strapiSource = new URLSearchParams();
            strapiSource.set('documentId', documentId);
            strapiSource.set('type', type);
            strapiSource.set('path', path);
            if (model) {
                strapiSource.set('model', model);
            }
            if (kind) {
                strapiSource.set('kind', kind);
            }
            if (locale) {
                strapiSource.set('locale', locale);
            }
            const encoded = stega.vercelStegaCombine(text, {
                strapiSource: strapiSource.toString()
            }, false);
            return encoded;
        },
        async encodeEntry ({ data, schema }) {
            if (!isObject(data) || data === undefined) {
                return data;
            }
            return strapiUtils.traverseEntity(({ key, value, attribute, schema, path, parent }, { set })=>{
                if (!attribute || EXCLUDED_FIELDS.includes(key)) {
                    return;
                }
                if (ENCODABLE_TYPES.includes(attribute.type) && typeof value === 'string') {
                    // For inner fields of a multi-media field's items (e.g. `medias.0.url`),
                    // drop the array index so all items share the same encoded path. The
                    // preview groups them under one highlight and opens the multi-media
                    // input as a single field, matching the side editor.
                    const parentAttr = parent?.attribute;
                    const isInsideMultiMedia = parentAttr?.type === 'media' && parentAttr.multiple === true;
                    const encodedPath = isInsideMultiMedia && parent?.path?.rawWithIndices ? `${parent.path.rawWithIndices}.${key}` : path.rawWithIndices;
                    set(key, this.encodeField(value, {
                        path: encodedPath,
                        type: attribute.type,
                        kind: schema.kind,
                        model: schema.uid,
                        locale: data.locale,
                        documentId: data.documentId
                    }));
                }
            }, {
                schema,
                getModel: (uid)=>strapi.getModel(uid)
            }, data);
        },
        async encodeSourceMaps ({ data, schema }) {
            try {
                if (Array.isArray(data)) {
                    return await Promise.all(data.map((item)=>this.encodeSourceMaps({
                            data: item,
                            schema
                        })));
                }
                if (!isObject(data)) {
                    return data;
                }
                return await this.encodeEntry({
                    data,
                    schema
                });
            } catch (error) {
                strapi.log.error('Error encoding source maps:', error);
                return data;
            }
        }
    };
};

exports.createContentSourceMapsService = createContentSourceMapsService;
