// @strapi/content-manager/dist/admin/pages/formatComponentConfigurationEditLayout.js
'use strict';

var useDocumentLayout = require('../hooks/useDocumentLayout.js');

/**
 * Formats API configuration + schemas into the shape expected by the component
 * "Configure the view" form. Matches the ListView / content-type path: nested
 * `convertEditLayoutToFieldLayouts` calls must receive component configuration
 * and full component schemas so `getMainField` can resolve fields inside nested
 * component attributes (see #25509).
 */ const formatComponentConfigurationEditLayout = (data, { schema, components })=>{
    const editAttributes = useDocumentLayout.convertEditLayoutToFieldLayouts(data.component.layouts.edit, schema?.attributes, data.component.metadatas, {
        configurations: data.components,
        schemas: components
    });
    const componentEditAttributes = Object.entries(data.components).reduce((acc, [uid, configuration])=>{
        const componentSchema = components[uid];
        if (!componentSchema) {
            return acc;
        }
        acc[uid] = {
            layout: useDocumentLayout.convertEditLayoutToFieldLayouts(configuration.layouts.edit, componentSchema.attributes, configuration.metadatas, {
                configurations: data.components,
                schemas: components
            }),
            settings: {
                ...configuration.settings,
                icon: componentSchema.info.icon,
                displayName: componentSchema.info.displayName
            }
        };
        return acc;
    }, {});
    const editMetadatas = Object.entries(data.component.metadatas).reduce((acc, [attribute, metadata])=>{
        return {
            ...acc,
            [attribute]: metadata.edit
        };
    }, {});
    return {
        layout: [
            editAttributes
        ],
        components: componentEditAttributes,
        metadatas: editMetadatas,
        options: {
            ...schema?.options,
            ...schema?.pluginOptions
        },
        settings: {
            ...data.component.settings,
            displayName: schema?.info.displayName
        }
    };
};

exports.formatComponentConfigurationEditLayout = formatComponentConfigurationEditLayout;
