// @strapi/review-workflows/dist/admin/routes/content-manager/model/components/AssigneeFilter.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var users = require('../../../../utils/users.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const PAGE_SIZE = 10;
const AssigneeFilter = ({ name })=>{
    const [pageSize, setPageSize] = React__namespace.useState(PAGE_SIZE);
    const [hasOpened, setHasOpened] = React__namespace.useState(false);
    const { formatMessage } = reactIntl.useIntl();
    const field = strapiAdmin.useField(name);
    const shouldFetch = hasOpened || Boolean(field.value);
    const { data, isLoading } = strapiAdmin.useAdminUsers({
        pageSize
    }, {
        skip: !shouldFetch
    });
    const handleChange = (value)=>{
        setPageSize(PAGE_SIZE);
        field.onChange(name, value);
    };
    const handleOpenChange = (isOpen)=>{
        if (isOpen && !hasOpened) {
            setHasOpened(true);
        }
    };
    const { pageCount = 1, page = 1 } = data?.pagination ?? {};
    const hasMoreItems = page < pageCount;
    const handleLoadMore = ()=>{
        if (!hasMoreItems) {
            return;
        }
        setPageSize((prev)=>prev + PAGE_SIZE);
    };
    const options = React__namespace.useMemo(()=>(data?.users ?? []).map((user)=>({
                id: user.id,
                label: users.getDisplayName(user)
            })), [
        data?.users
    ]);
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Combobox, {
        value: field.value,
        "aria-label": formatMessage({
            id: 'content-manager.components.Filters.usersSelect.label',
            defaultMessage: 'Search and select an user to filter'
        }),
        onChange: handleChange,
        onOpenChange: handleOpenChange,
        loading: isLoading,
        onLoadMore: handleLoadMore,
        hasMoreItems: hasMoreItems,
        children: options.map((option)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.ComboboxOption, {
                value: option.id.toString(),
                children: option.label
            }, option.id))
    });
};

exports.AssigneeFilter = AssigneeFilter;
