// @strapi/review-workflows/dist/admin/routes/content-manager/model/components/AssigneeFilter.mjs
import { jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { useField, useAdminUsers } from '@strapi/admin/strapi-admin';
import { Combobox, ComboboxOption } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { getDisplayName } from '../../../../utils/users.mjs';

const PAGE_SIZE = 10;
const AssigneeFilter = ({ name })=>{
    const [pageSize, setPageSize] = React.useState(PAGE_SIZE);
    const [hasOpened, setHasOpened] = React.useState(false);
    const { formatMessage } = useIntl();
    const field = useField(name);
    const shouldFetch = hasOpened || Boolean(field.value);
    const { data, isLoading } = useAdminUsers({
        pageSize
    }, {
        skip: !shouldFetch
    });
    const handleChange = (value)=>{
        setPageSize(PAGE_SIZE);
        field.onChange(name, value);
    };
    const handleOpenChange = (isOpen)=>{
        if (isOpen && !hasOpened) {
            setHasOpened(true);
        }
    };
    const { pageCount = 1, page = 1 } = data?.pagination ?? {};
    const hasMoreItems = page < pageCount;
    const handleLoadMore = ()=>{
        if (!hasMoreItems) {
            return;
        }
        setPageSize((prev)=>prev + PAGE_SIZE);
    };
    const options = React.useMemo(()=>(data?.users ?? []).map((user)=>({
                id: user.id,
                label: getDisplayName(user)
            })), [
        data?.users
    ]);
    return /*#__PURE__*/ jsx(Combobox, {
        value: field.value,
        "aria-label": formatMessage({
            id: 'content-manager.components.Filters.usersSelect.label',
            defaultMessage: 'Search and select an user to filter'
        }),
        onChange: handleChange,
        onOpenChange: handleOpenChange,
        loading: isLoading,
        onLoadMore: handleLoadMore,
        hasMoreItems: hasMoreItems,
        children: options.map((option)=>/*#__PURE__*/ jsx(ComboboxOption, {
                value: option.id.toString(),
                children: option.label
            }, option.id))
    });
};

export { AssigneeFilter };
