// @strapi/review-workflows/dist/admin/routes/content-manager/model/components/StageFilter.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { useField } from '@strapi/admin/strapi-admin';
import { SingleSelect, SingleSelectOption, Flex, Typography, Loader } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { useParams } from 'react-router-dom';
import { getStageColorByHex } from '../../../../utils/colors.mjs';
import { useReviewWorkflows } from '../../../settings/hooks/useReviewWorkflows.mjs';

const StageFilter = (props)=>{
    const { slug: model } = useParams();
    const { formatMessage } = useIntl();
    const { workflows, isLoading } = useReviewWorkflows({
        filters: {
            contentTypes: model
        }
    });
    const field = useField(props.name);
    const [workflow] = workflows ?? [];
    return /*#__PURE__*/ jsx(SingleSelect, {
        "aria-label": formatMessage({
            id: 'content-manager.components.Filters.reviewWorkflows.label',
            defaultMessage: 'Search and select an workflow stage to filter'
        }),
        value: field.value,
        onChange: (value)=>{
            field.onChange(props.name, value);
        },
        loading: isLoading,
        // @ts-expect-error – DS type error with SingleSelect['customizeContent']
        customizeContent: ()=>/*#__PURE__*/ jsxs(Flex, {
                tag: "span",
                justifyContent: "space-between",
                alignItems: "center",
                width: "100%",
                children: [
                    /*#__PURE__*/ jsx(Typography, {
                        textColor: "neutral800",
                        ellipsis: true,
                        children: field.value
                    }),
                    isLoading ? /*#__PURE__*/ jsx(Loader, {
                        small: true,
                        style: {
                            display: 'flex'
                        }
                    }) : null
                ]
            }),
        children: (workflow?.stages ?? []).map(({ id, color, name })=>{
            const { themeColorName } = getStageColorByHex(color) ?? {};
            return /*#__PURE__*/ jsx(SingleSelectOption, {
                startIcon: /*#__PURE__*/ jsx(Flex, {
                    height: 2,
                    background: color,
                    borderColor: themeColorName === 'neutral0' ? 'neutral150' : undefined,
                    hasRadius: true,
                    shrink: 0,
                    width: 2
                }),
                value: name,
                children: name
            }, id);
        })
    });
};

export { StageFilter };
