// @strapi/review-workflows/dist/admin/routes/content-manager/model/components/StageFilter.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var colors = require('../../../../utils/colors.js');
var useReviewWorkflows = require('../../../settings/hooks/useReviewWorkflows.js');

const StageFilter = (props)=>{
    const { slug: model } = reactRouterDom.useParams();
    const { formatMessage } = reactIntl.useIntl();
    const { workflows, isLoading } = useReviewWorkflows.useReviewWorkflows({
        filters: {
            contentTypes: model
        }
    });
    const field = strapiAdmin.useField(props.name);
    const [workflow] = workflows ?? [];
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.SingleSelect, {
        "aria-label": formatMessage({
            id: 'content-manager.components.Filters.reviewWorkflows.label',
            defaultMessage: 'Search and select an workflow stage to filter'
        }),
        value: field.value,
        onChange: (value)=>{
            field.onChange(props.name, value);
        },
        loading: isLoading,
        // @ts-expect-error – DS type error with SingleSelect['customizeContent']
        customizeContent: ()=>/*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                tag: "span",
                justifyContent: "space-between",
                alignItems: "center",
                width: "100%",
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                        textColor: "neutral800",
                        ellipsis: true,
                        children: field.value
                    }),
                    isLoading ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.Loader, {
                        small: true,
                        style: {
                            display: 'flex'
                        }
                    }) : null
                ]
            }),
        children: (workflow?.stages ?? []).map(({ id, color, name })=>{
            const { themeColorName } = colors.getStageColorByHex(color) ?? {};
            return /*#__PURE__*/ jsxRuntime.jsx(designSystem.SingleSelectOption, {
                startIcon: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                    height: 2,
                    background: color,
                    borderColor: themeColorName === 'neutral0' ? 'neutral150' : undefined,
                    hasRadius: true,
                    shrink: 0,
                    width: 2
                }),
                value: name,
                children: name
            }, id);
        })
    });
};

exports.StageFilter = StageFilter;
