// @strapi/admin/dist/server/server/src/strategies/api-token-utils.js
'use strict';

var fp = require('lodash/fp');
var dateFns = require('date-fns');
var utils = require('@strapi/utils');

const { UnauthorizedError } = utils.errors;
const extractToken = (ctx)=>{
    if (ctx.request && ctx.request.header && ctx.request.header.authorization) {
        const parts = ctx.request.header.authorization.split(/\s+/);
        if (parts[0].toLowerCase() !== 'bearer' || parts.length !== 2) {
            return null;
        }
        return parts[1];
    }
    return null;
};
const checkExpiry = (apiToken)=>{
    if (!fp.isNil(apiToken.expiresAt)) {
        const expirationDate = new Date(apiToken.expiresAt);
        if (expirationDate < new Date()) {
            return new UnauthorizedError('Token expired');
        }
    }
    return null;
};
const updateLastUsedAt = async (apiToken)=>{
    const currentDate = new Date();
    if (!fp.isNil(apiToken.lastUsedAt)) {
        const hoursSinceLastUsed = dateFns.differenceInHours(currentDate, dateFns.parseISO(apiToken.lastUsedAt));
        if (hoursSinceLastUsed >= 1) {
            await strapi.db.query('admin::api-token').update({
                where: {
                    id: apiToken.id
                },
                data: {
                    lastUsedAt: currentDate
                }
            });
        }
    } else {
        await strapi.db.query('admin::api-token').update({
            where: {
                id: apiToken.id
            },
            data: {
                lastUsedAt: currentDate
            }
        });
    }
};

exports.checkExpiry = checkExpiry;
exports.extractToken = extractToken;
exports.updateLastUsedAt = updateLastUsedAt;
