// @strapi/admin/dist/server/server/src/strategies/content-api-token.mjs
import { castArray } from 'lodash/fp';
import { errors } from '@strapi/utils';
import constants from '../services/constants.mjs';
import { getService } from '../utils/index.mjs';
import { checkExpiry, extractToken, updateLastUsedAt } from './api-token-utils.mjs';
import '@strapi/types';

const { UnauthorizedError, ForbiddenError } = errors;
const isReadScope = (scope)=>scope.endsWith('find') || scope.endsWith('findOne');
/**
 * Authenticate a content-api token. Rejects tokens with kind !== 'content-api'.
 */ const authenticate = async (ctx)=>{
    const apiTokenService = getService('api-token-content-api');
    const token = extractToken(ctx);
    if (token === null) {
        return {
            authenticated: false
        };
    }
    const apiToken = await apiTokenService.getByAccessKey(apiTokenService.hash(token));
    if (apiToken === null || apiToken === undefined) {
        return {
            authenticated: false
        };
    }
    // Defensive kind check — only handle content-api tokens.
    // null kind is allowed: tokens created before the kind field was introduced are implicitly content-api.
    if (apiToken.kind !== 'content-api' && apiToken.kind !== null) {
        return {
            authenticated: false
        };
    }
    const expiryError = checkExpiry(apiToken);
    if (expiryError !== null) {
        return {
            authenticated: false,
            error: expiryError
        };
    }
    await updateLastUsedAt(apiToken);
    if (apiToken.type === constants.API_TOKEN_TYPE.CUSTOM) {
        const ability = await strapi.contentAPI.permissions.engine.generateAbility(apiToken.permissions.map((action)=>({
                action
            })));
        return {
            authenticated: true,
            ability,
            credentials: apiToken
        };
    }
    return {
        authenticated: true,
        credentials: apiToken
    };
};
/**
 * Verify the token has the required abilities for the requested scope.
 *
 * @type {import('.').VerifyFunction}
 */ const verify = (auth, config)=>{
    const { credentials: apiToken, ability } = auth;
    if (apiToken === null || apiToken === undefined) {
        throw new UnauthorizedError('Token not found');
    }
    const expiryError = checkExpiry(apiToken);
    if (expiryError !== null) {
        throw expiryError;
    }
    // Full access
    if (apiToken.type === constants.API_TOKEN_TYPE.FULL_ACCESS) {
        return;
    }
    // Read only
    if (apiToken.type === constants.API_TOKEN_TYPE.READ_ONLY) {
        const scopes = castArray(config.scope);
        if (config.scope && scopes.every(isReadScope)) {
            return;
        }
    } else if (apiToken.type === constants.API_TOKEN_TYPE.CUSTOM) {
        if (ability === null || ability === undefined) {
            throw new ForbiddenError();
        }
        const scopes = castArray(config.scope);
        const isAllowed = scopes.every((scope)=>ability.can(scope));
        if (isAllowed === true) {
            return;
        }
    }
    throw new ForbiddenError();
};
var contentApiTokenAuthStrategy = {
    name: 'content-api-token',
    authenticate,
    verify
};

export { authenticate, contentApiTokenAuthStrategy as default, verify };
