// @strapi/strapi/dist/src/cli/commands/admin/block-user.js
'use strict';

var _ = require('lodash');
var commander = require('commander');
var core = require('@strapi/core');
var helpers = require('../../utils/helpers.js');
var getInquirer = require('../../utils/get-inquirer.js');

const promptQuestions = [
    {
        type: 'input',
        name: 'email',
        message: 'User email?'
    },
    {
        type: 'input',
        name: 'block',
        message: 'User Blocked?'
    }
];
async function setBlock({ email, block }) {
    const appContext = await core.compileStrapi();
    const app = await core.createStrapi(appContext).load();
    const cleanEmail = email?.trim().toLowerCase();
    const cleanBlock = block?.trim().toLowerCase();
    if (cleanBlock !== 'true' && cleanBlock !== 'false') {
        console.error('Invalid block status. Use "true" or "false".');
        process.exit(1);
    }
    const blocked = cleanBlock === 'true';
    const user = await app.admin.services.user.findOneByEmail(cleanEmail);
    if (!user) {
        console.error(`User with email "${cleanEmail}" does not exist`);
        process.exit(1);
    }
    try {
        await app.admin.services.user.updateById(user.id, {
            blocked
        });
    } catch (err) {
        console.error(err.message);
        process.exit(1);
    }
    console.log(`Successfully set ${cleanEmail} block status to ${cleanBlock}`);
    process.exit(0);
}
/**
 * Change a user's block status
 */ const action = async (cmdOptions = {})=>{
    const { email, block } = cmdOptions;
    if (_.isEmpty(email) && _.isEmpty(block) && process.stdin.isTTY) {
        const inquirer = await getInquirer.getInquirer();
        const inquiry = await inquirer.prompt(promptQuestions);
        return setBlock(inquiry);
    }
    if (_.isEmpty(email) || _.isEmpty(block)) {
        console.error('Missing required options `email` or `block`');
        process.exit(1);
    }
    return setBlock({
        email,
        block
    });
};
/**
 * `$ strapi admin:block-user`
 */ const command = ()=>{
    return commander.createCommand('admin:block-user').alias('admin:block').description("Set a user's block status").option('-e, --email <email>', 'The user email').option('-b, --block <true/false>', 'The user block status').action(helpers.runAction('admin:block-user', action));
};

exports.action = action;
exports.command = command;
