// @strapi/content-manager/dist/admin/pages/ListView/components/Filters.mjs
import { jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { Filters, useTracking, useAuth, useQueryParams, useStrapiApp, useAdminUsers, useField } from '@strapi/admin/strapi-admin';
import { useCollator, Combobox, ComboboxOption } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { CREATOR_FIELDS } from '../../../constants/attributes.mjs';
import { HOOKS } from '../../../constants/hooks.mjs';
import { useContentTypeSchema } from '../../../hooks/useContentTypeSchema.mjs';
import { useDebounce } from '../../../hooks/useDebounce.mjs';
import { useGetContentTypeConfigurationQuery } from '../../../services/contentTypes.mjs';
import { getMainField } from '../../../utils/attributes.mjs';
import { getTranslation } from '../../../utils/translations.mjs';
import { getDisplayName } from '../../../utils/users.mjs';

const { INJECT_LIST_VIEW_FILTERS } = HOOKS;
/**
 * If new attributes are added, this list needs to be updated.
 */ const NOT_ALLOWED_FILTERS = [
    'json',
    'component',
    'media',
    'richtext',
    'dynamiczone',
    'password',
    'blocks'
];
const DEFAULT_ALLOWED_FILTERS = [
    'createdAt',
    'updatedAt'
];
const USER_FILTER_ATTRIBUTES = [
    ...CREATOR_FIELDS,
    'strapi_assignee'
];
const Root = ({ disabled, schema, layout, children })=>{
    const { attributes, uid: model, options } = schema;
    const { formatMessage, locale } = useIntl();
    const { trackUsage } = useTracking();
    const allPermissions = useAuth('FiltersImpl', (state)=>state.permissions);
    const [{ query }] = useQueryParams();
    const { schemas } = useContentTypeSchema();
    const runHookWaterfall = useStrapiApp('FiltersImpl', ({ runHookWaterfall })=>runHookWaterfall);
    const canReadAdminUsers = React.useMemo(()=>allPermissions.filter((permission)=>permission.action === 'admin::users.read' && permission.subject === null).length > 0, [
        allPermissions
    ]);
    const selectedUserIds = (query?.filters?.$and ?? []).reduce((acc, filter)=>{
        const [key, value] = Object.entries(filter)[0];
        if (typeof value.id !== 'object') {
            return acc;
        }
        const id = value.id.$eq || value.id.$ne;
        // Check if the attribute is a relation to admin::user
        const attribute = attributes[key];
        const isAdminUserRelation = attribute?.type === 'relation' && 'target' in attribute && attribute.target === 'admin::user';
        if (id && (isAdminUserRelation || USER_FILTER_ATTRIBUTES.includes(key)) && !acc.includes(id)) {
            acc.push(id);
        }
        return acc;
    }, []);
    const { data: userData, isLoading: isLoadingAdminUsers } = useAdminUsers({
        filters: {
            id: {
                $in: selectedUserIds
            }
        }
    }, {
        // fetch the list of admin users only if the filter contains users and the
        // current user has permissions to display users
        skip: selectedUserIds.length === 0 || !canReadAdminUsers
    });
    const { users = [] } = userData ?? {};
    const { metadata } = useGetContentTypeConfigurationQuery(model, {
        selectFromResult: ({ data })=>({
                metadata: data?.contentType.metadatas ?? {}
            })
    });
    const formatter = useCollator(locale, {
        sensitivity: 'base'
    });
    const displayedFilters = React.useMemo(()=>{
        const [{ properties: { fields = [] } = {
            fields: []
        } }] = allPermissions.filter((permission)=>permission.action === 'plugin::content-manager.explorer.read' && permission.subject === model);
        const allowedFields = fields.filter((field)=>{
            const attribute = attributes[field] ?? {};
            return attribute.type && !NOT_ALLOWED_FILTERS.includes(attribute.type);
        });
        const baseFilters = [
            'id',
            'documentId',
            ...allowedFields,
            ...DEFAULT_ALLOWED_FILTERS,
            ...canReadAdminUsers ? CREATOR_FIELDS : [],
            ...options?.draftAndPublish === true ? [
                '__status'
            ] : []
        ].map((name)=>{
            if (name === '__status') {
                return {
                    name: '__status',
                    type: 'enumeration',
                    label: formatMessage({
                        id: getTranslation('containers.list.filters.status'),
                        defaultMessage: 'Status'
                    }),
                    operators: [
                        {
                            label: formatMessage({
                                id: 'components.FilterOptions.FILTER_TYPES.$eq',
                                defaultMessage: 'is'
                            }),
                            value: '$eq'
                        }
                    ],
                    options: [
                        {
                            label: formatMessage({
                                id: getTranslation('containers.List.statusFilter.draft'),
                                defaultMessage: 'Draft (never published)'
                            }),
                            value: 'draft'
                        },
                        {
                            label: formatMessage({
                                id: getTranslation('containers.List.statusFilter.published'),
                                defaultMessage: 'Published (all)'
                            }),
                            value: 'published'
                        },
                        {
                            label: formatMessage({
                                id: getTranslation('containers.List.statusFilter.publishedModified'),
                                defaultMessage: 'Published (modified)'
                            }),
                            value: 'published-modified'
                        },
                        {
                            label: formatMessage({
                                id: getTranslation('containers.List.statusFilter.publishedUnmodified'),
                                defaultMessage: 'Published (unmodified)'
                            }),
                            value: 'published-unmodified'
                        }
                    ]
                };
            }
            const attribute = attributes[name];
            if (NOT_ALLOWED_FILTERS.includes(attribute.type)) {
                return null;
            }
            const { mainField: mainFieldName = '', label } = metadata[name].list;
            let filter = {
                name,
                label: label ?? '',
                mainField: getMainField(attribute, mainFieldName, {
                    schemas,
                    components: {}
                }),
                type: attribute.type
            };
            if (attribute.type === 'relation' && 'target' in attribute && attribute.target === 'admin::user') {
                filter = {
                    ...filter,
                    input: AdminUsersFilter,
                    options: users.map((user)=>({
                            label: getDisplayName(user),
                            value: user.id.toString()
                        })),
                    operators: [
                        {
                            label: formatMessage({
                                id: 'components.FilterOptions.FILTER_TYPES.$eq',
                                defaultMessage: 'is'
                            }),
                            value: '$eq'
                        },
                        {
                            label: formatMessage({
                                id: 'components.FilterOptions.FILTER_TYPES.$ne',
                                defaultMessage: 'is not'
                            }),
                            value: '$ne'
                        }
                    ],
                    mainField: {
                        name: 'id',
                        type: 'integer'
                    }
                };
            }
            if (attribute.type === 'enumeration') {
                filter = {
                    ...filter,
                    options: attribute.enum.map((value)=>({
                            label: value,
                            value
                        }))
                };
            }
            return filter;
        }).filter(Boolean);
        // Let plugins inject their own filters.
        // `layout.options` is a merge of schema/pluginOptions/contentType options —
        // mirrors the column-injection hook contract.
        const { displayedFilters: extendedFilters } = runHookWaterfall(INJECT_LIST_VIEW_FILTERS, {
            displayedFilters: baseFilters,
            layout
        });
        const resolveLabel = (label)=>typeof label === 'string' ? label : formatMessage(label);
        const userOptions = users.map((user)=>({
                label: getDisplayName(user),
                value: user.id.toString()
            }));
        const formatted = extendedFilters.map((filter)=>{
            const next = {
                ...filter,
                label: resolveLabel(filter.label),
                operators: filter.operators?.map((op)=>({
                        ...op,
                        label: resolveLabel(op.label)
                    }))
            };
            // User-typed filters need user options so the chip can render the display name instead of the id.
            if (USER_FILTER_ATTRIBUTES.includes(filter.name) && !next.options?.length) {
                next.options = userOptions;
            }
            return next;
        });
        return formatted.toSorted((a, b)=>formatter.compare(a.label, b.label));
    }, [
        allPermissions,
        canReadAdminUsers,
        model,
        attributes,
        metadata,
        schemas,
        layout,
        users,
        formatMessage,
        formatter,
        options?.draftAndPublish,
        runHookWaterfall
    ]);
    const onOpenChange = (isOpen)=>{
        if (isOpen) {
            trackUsage('willFilterEntries');
        }
    };
    const handleFilterChange = (data)=>{
        const attribute = attributes[data.name];
        if (attribute) {
            trackUsage('didFilterEntries', {
                useRelation: attribute.type === 'relation'
            });
        }
    };
    return /*#__PURE__*/ jsx(Filters.Root, {
        disabled: disabled,
        options: displayedFilters,
        onOpenChange: onOpenChange,
        onChange: handleFilterChange,
        children: children
    });
};
/* -------------------------------------------------------------------------------------------------
 * AdminUsersFilter
 * -----------------------------------------------------------------------------------------------*/ const AdminUsersFilter = ({ name })=>{
    const [pageSize, setPageSize] = React.useState(10);
    const [search, setSearch] = React.useState('');
    const { formatMessage } = useIntl();
    const debouncedSearch = useDebounce(search, 300);
    const { data, isLoading } = useAdminUsers({
        pageSize,
        _q: debouncedSearch
    });
    const field = useField(name);
    const handleOpenChange = (isOpen)=>{
        if (!isOpen) {
            setPageSize(10);
        }
    };
    const { users = [], pagination } = data ?? {};
    const { pageCount = 1, page = 1 } = pagination ?? {};
    return /*#__PURE__*/ jsx(Combobox, {
        value: field.value,
        "aria-label": formatMessage({
            id: 'content-manager.components.Filters.usersSelect.label',
            defaultMessage: 'Search and select a user to filter'
        }),
        onOpenChange: handleOpenChange,
        onChange: (value)=>field.onChange(name, value),
        loading: isLoading,
        onLoadMore: ()=>setPageSize(pageSize + 10),
        hasMoreItems: page < pageCount,
        onInputChange: (e)=>{
            setSearch(e.currentTarget.value);
        },
        children: users.map((user)=>{
            return /*#__PURE__*/ jsx(ComboboxOption, {
                value: user.id.toString(),
                children: getDisplayName(user)
            }, user.id);
        })
    });
};
const listViewFilters = {
    Root,
    Trigger: Filters.Trigger,
    Popover: Filters.Popover,
    List: Filters.List
};

export { listViewFilters };
