// @strapi/admin/dist/server/server/src/controllers/admin-token.mjs
import { errors, strings } from '@strapi/utils';
import { has, trim } from 'lodash/fp';
import { getService } from '../utils/index.mjs';
import constants from '../services/constants.mjs';
import { validateAdminTokenUpdateInput, validateAdminTokenCreationInput } from '../validation/admin-tokens.mjs';

const { ApplicationError } = errors;
// ---------------------------------------------------------------------------
// Access-control helpers
// ---------------------------------------------------------------------------
const isSuperAdmin = (user)=>user.roles.some((r)=>r.code === constants.SUPER_ADMIN_CODE) === true;
const getOwnerId = (token)=>{
    const owner = token.adminUserOwner;
    return String(typeof owner === 'object' ? owner.id : owner);
};
/** Returns true when user is the recorded owner of an admin token. */ const isTokenOwner = (user, token)=>getOwnerId(token) === String(user.id);
/** Owner OR super-admin can manage an admin token (read metadata, update…). */ const canAccessAdminToken = (user, token)=>isTokenOwner(user, token) || isSuperAdmin(user);
// ---------------------------------------------------------------------------
// Controller
// ---------------------------------------------------------------------------
var adminToken = {
    // -------------------------------------------------------------------------
    // Create
    // -------------------------------------------------------------------------
    async create (ctx) {
        const { body } = ctx.request;
        const apiTokenService = getService('api-token-admin');
        if (body.type !== undefined) {
            return ctx.badRequest('Type is not allowed for admin tokens');
        }
        if (body.permissions !== undefined) {
            return ctx.badRequest('Permissions are not allowed for admin tokens');
        }
        const attributes = {
            kind: 'admin',
            name: trim(body.name),
            description: trim(body.description),
            adminPermissions: body.adminPermissions,
            lifespan: body.lifespan,
            adminUserOwner: ctx.state.user.id
        };
        await validateAdminTokenCreationInput(attributes);
        const alreadyExists = await apiTokenService.exists({
            name: attributes.name
        });
        if (alreadyExists) {
            throw new ApplicationError('Name already taken');
        }
        const apiToken = await apiTokenService.create(attributes, ctx.state.user);
        ctx.created({
            data: apiToken
        });
    },
    // -------------------------------------------------------------------------
    // Regenerate — owner-only, super-admin does NOT bypass
    // -------------------------------------------------------------------------
    async regenerate (ctx) {
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-admin');
        const token = await apiTokenService.getById(id);
        if (!token) {
            ctx.notFound('API Token not found');
            return;
        }
        if (!isTokenOwner(ctx.state.user, token)) {
            return ctx.forbidden();
        }
        const accessToken = await apiTokenService.regenerate(id);
        ctx.created({
            data: accessToken
        });
    },
    // -------------------------------------------------------------------------
    // List — always filtered to kind: 'admin'
    // -------------------------------------------------------------------------
    async list (ctx) {
        const apiTokenService = getService('api-token-admin');
        const apiTokens = await apiTokenService.list(ctx.state.user);
        ctx.send({
            data: apiTokens
        });
    },
    // -------------------------------------------------------------------------
    // Revoke
    // -------------------------------------------------------------------------
    async revoke (ctx) {
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-admin');
        const existingToken = await apiTokenService.getById(id);
        if (existingToken === null) {
            return ctx.notFound('API Token not found');
        }
        if (canAccessAdminToken(ctx.state.user, existingToken) === false) {
            return ctx.forbidden();
        }
        const apiToken = await apiTokenService.revoke(id);
        ctx.deleted({
            data: apiToken
        });
    },
    // -------------------------------------------------------------------------
    // Get — key exposed only to owner
    // -------------------------------------------------------------------------
    async get (ctx) {
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-admin');
        const token = await apiTokenService.getById(id);
        if (!token) {
            ctx.notFound('API Token not found');
            return;
        }
        if (canAccessAdminToken(ctx.state.user, token) === false) {
            ctx.notFound('API Token not found');
            return;
        }
        if (isTokenOwner(ctx.state.user, token)) {
            const withKey = await apiTokenService.getById(id, {
                includeDecryptedKey: true
            });
            ctx.send({
                data: withKey ?? token
            });
            return;
        }
        ctx.send({
            data: token
        });
    },
    // -------------------------------------------------------------------------
    // Update — owner or super-admin only
    // -------------------------------------------------------------------------
    async update (ctx) {
        const { body } = ctx.request;
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-admin');
        const mutableBody = body;
        if (has('name', mutableBody)) {
            mutableBody.name = trim(body.name ?? '');
        }
        if (has('description', mutableBody) || mutableBody.description === null) {
            mutableBody.description = trim(body.description ?? '');
        }
        await validateAdminTokenUpdateInput(body);
        const existingToken = await apiTokenService.getById(id);
        if (!existingToken) {
            return ctx.notFound('API Token not found');
        }
        if (has('name', body)) {
            const nameAlreadyTaken = await apiTokenService.getByName(body.name);
            if (nameAlreadyTaken !== null && !strings.isEqual(nameAlreadyTaken.id, id)) {
                throw new ApplicationError('Name already taken');
            }
        }
        if (!canAccessAdminToken(ctx.state.user, existingToken)) {
            return ctx.forbidden();
        }
        const apiToken = await apiTokenService.update(id, body);
        ctx.send({
            data: apiToken
        });
    },
    // -------------------------------------------------------------------------
    // Owner permissions — effective permissions of the token owner
    // -------------------------------------------------------------------------
    async getOwnerPermissions (ctx) {
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-admin');
        const permissionService = getService('permission');
        const userService = getService('user');
        const token = await apiTokenService.getById(id);
        if (!token) {
            return ctx.notFound('apiToken.notFound');
        }
        if (!canAccessAdminToken(ctx.state.user, token)) {
            return ctx.forbidden();
        }
        const ownerId = getOwnerId(token);
        const ownerUser = await userService.findOne(ownerId);
        if (!ownerUser) {
            return ctx.notFound('owner.notFound');
        }
        const ownerPermissions = await permissionService.findUserPermissions(ownerUser);
        const sanitizedPermissions = ownerPermissions.map(permissionService.sanitizePermission);
        // @ts-expect-error - transform response type to sanitized permission
        ctx.body = {
            data: sanitizedPermissions
        };
    }
};

export { adminToken as default };
