// @strapi/admin/dist/server/server/src/controllers/api-token.mjs
import { errors, strings } from '@strapi/utils';
import { has, trim } from 'lodash/fp';
import { getService } from '../utils/index.mjs';
import { validateApiTokenUpdateInput, validateApiTokenCreationInput } from '../validation/api-tokens.mjs';

const { ApplicationError } = errors;
var apiToken = {
    // -------------------------------------------------------------------------
    // Create
    // -------------------------------------------------------------------------
    async create (ctx) {
        const { body } = ctx.request;
        const apiTokenService = getService('api-token-content-api');
        const attributes = {
            kind: 'content-api',
            name: trim(body.name),
            description: trim(body.description),
            type: body.type,
            permissions: body.permissions,
            lifespan: body.lifespan
        };
        await validateApiTokenCreationInput(attributes);
        const alreadyExists = await apiTokenService.exists({
            name: attributes.name
        });
        if (alreadyExists) {
            throw new ApplicationError('Name already taken');
        }
        const apiToken = await apiTokenService.create(attributes, ctx.state.user);
        ctx.created({
            data: apiToken
        });
    },
    // -------------------------------------------------------------------------
    // Regenerate
    // -------------------------------------------------------------------------
    async regenerate (ctx) {
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-content-api');
        const token = await apiTokenService.getById(id);
        if (!token) {
            ctx.notFound('API Token not found');
            return;
        }
        const accessToken = await apiTokenService.regenerate(id);
        ctx.created({
            data: accessToken
        });
    },
    // -------------------------------------------------------------------------
    // List — always content-api
    // -------------------------------------------------------------------------
    async list (ctx) {
        const apiTokenService = getService('api-token-content-api');
        const apiTokens = await apiTokenService.list(ctx.state.user);
        ctx.send({
            data: apiTokens
        });
    },
    // -------------------------------------------------------------------------
    // Revoke
    // -------------------------------------------------------------------------
    async revoke (ctx) {
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-content-api');
        const apiToken = await apiTokenService.revoke(id);
        ctx.deleted({
            data: apiToken
        });
    },
    // -------------------------------------------------------------------------
    // Get — always expose the decrypted key (content-api back-compat)
    // -------------------------------------------------------------------------
    async get (ctx) {
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-content-api');
        const token = await apiTokenService.getById(id, {
            includeDecryptedKey: true
        });
        if (!token) {
            ctx.notFound('API Token not found');
            return;
        }
        ctx.send({
            data: token
        });
    },
    // -------------------------------------------------------------------------
    // Update
    // -------------------------------------------------------------------------
    async update (ctx) {
        const { body } = ctx.request;
        const { id } = ctx.params;
        const apiTokenService = getService('api-token-content-api');
        const mutableBody = body;
        if (has('name', mutableBody)) {
            mutableBody.name = trim(body.name ?? '');
        }
        if (has('description', mutableBody) || mutableBody.description === null) {
            mutableBody.description = trim(body.description ?? '');
        }
        await validateApiTokenUpdateInput(body);
        const existingToken = await apiTokenService.getById(id);
        if (!existingToken) {
            return ctx.notFound('API Token not found');
        }
        if (has('name', body)) {
            const nameAlreadyTaken = await apiTokenService.getByName(body.name);
            if (nameAlreadyTaken !== null && !strings.isEqual(nameAlreadyTaken.id, id)) {
                throw new ApplicationError('Name already taken');
            }
        }
        const apiToken = await apiTokenService.update(id, body);
        ctx.send({
            data: apiToken
        });
    }
};

export { apiToken as default };
