// @strapi/admin/dist/admin/admin/src/pages/Settings/pages/Roles/hooks/usePermissionsDataManager.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactContext = require('@radix-ui/react-context');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

// eslint-disable-next-line check-file/filename-naming-convention
const [PermissionsDataManagerProviderRaw, usePermissionsDataManagerContext] = reactContext.createContext('PermissionsDataManager');
const usePermissionsDataManager = ()=>usePermissionsDataManagerContext('usePermissionsDataManager');
const PermissionsDataManagerProvider = ({ children, userPermissions, availableConditions, modifiedData, onChangeConditions, onChangeSimpleCheckbox, onChangeParentCheckbox, onChangeCollectionTypeLeftActionRowCheckbox, onChangeCollectionTypeGlobalActionCheckbox })=>{
    const checkUserHasPermission = React__namespace.useCallback((action, subject, field)=>{
        if (userPermissions === undefined) {
            return true;
        }
        const matchingPermission = userPermissions.find((perm)=>perm.action === action && perm.subject === subject);
        if (matchingPermission === undefined) {
            return false;
        }
        if (field === undefined) {
            return true;
        }
        const fields = matchingPermission.properties?.fields;
        if (fields === null || fields === undefined) {
            return true;
        }
        if (Array.isArray(fields) === false || fields.length === 0) {
            return false;
        }
        return fields.some((allowedField)=>allowedField === field || field.startsWith(`${allowedField}.`));
    }, [
        userPermissions
    ]);
    return /*#__PURE__*/ jsxRuntime.jsx(PermissionsDataManagerProviderRaw, {
        availableConditions: availableConditions,
        modifiedData: modifiedData,
        onChangeConditions: onChangeConditions,
        onChangeSimpleCheckbox: onChangeSimpleCheckbox,
        onChangeParentCheckbox: onChangeParentCheckbox,
        onChangeCollectionTypeLeftActionRowCheckbox: onChangeCollectionTypeLeftActionRowCheckbox,
        onChangeCollectionTypeGlobalActionCheckbox: onChangeCollectionTypeGlobalActionCheckbox,
        userPermissions: userPermissions,
        checkUserHasPermission: checkUserHasPermission,
        children: children
    });
};

exports.PermissionsDataManagerProvider = PermissionsDataManagerProvider;
exports.usePermissionsDataManager = usePermissionsDataManager;
