// @strapi/admin/dist/admin/admin/src/pages/Settings/pages/Roles/hooks/usePermissionsDataManager.mjs
import { jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { createContext } from '@radix-ui/react-context';

// eslint-disable-next-line check-file/filename-naming-convention
const [PermissionsDataManagerProviderRaw, usePermissionsDataManagerContext] = createContext('PermissionsDataManager');
const usePermissionsDataManager = ()=>usePermissionsDataManagerContext('usePermissionsDataManager');
const PermissionsDataManagerProvider = ({ children, userPermissions, availableConditions, modifiedData, onChangeConditions, onChangeSimpleCheckbox, onChangeParentCheckbox, onChangeCollectionTypeLeftActionRowCheckbox, onChangeCollectionTypeGlobalActionCheckbox })=>{
    const checkUserHasPermission = React.useCallback((action, subject, field)=>{
        if (userPermissions === undefined) {
            return true;
        }
        const matchingPermission = userPermissions.find((perm)=>perm.action === action && perm.subject === subject);
        if (matchingPermission === undefined) {
            return false;
        }
        if (field === undefined) {
            return true;
        }
        const fields = matchingPermission.properties?.fields;
        if (fields === null || fields === undefined) {
            return true;
        }
        if (Array.isArray(fields) === false || fields.length === 0) {
            return false;
        }
        return fields.some((allowedField)=>allowedField === field || field.startsWith(`${allowedField}.`));
    }, [
        userPermissions
    ]);
    return /*#__PURE__*/ jsx(PermissionsDataManagerProviderRaw, {
        availableConditions: availableConditions,
        modifiedData: modifiedData,
        onChangeConditions: onChangeConditions,
        onChangeSimpleCheckbox: onChangeSimpleCheckbox,
        onChangeParentCheckbox: onChangeParentCheckbox,
        onChangeCollectionTypeLeftActionRowCheckbox: onChangeCollectionTypeLeftActionRowCheckbox,
        onChangeCollectionTypeGlobalActionCheckbox: onChangeCollectionTypeGlobalActionCheckbox,
        userPermissions: userPermissions,
        checkUserHasPermission: checkUserHasPermission,
        children: children
    });
};

export { PermissionsDataManagerProvider, usePermissionsDataManager };
