// @strapi/content-manager/dist/server/controllers/validation/index.js
'use strict';

var _ = require('lodash');
var strapiUtils = require('@strapi/utils');
var zod = require('../../validation/zod.js');
require('@strapi/types');
require('../../services/utils/configuration/attributes.js');
require('../../services/utils/configuration/settings.js');

const { PaginationError, ValidationError } = strapiUtils.errors;
const TYPES = [
    'singleType',
    'collectionType'
];
/**
 * Validates type kind
 */ const kindSchema = strapiUtils.z.enum(TYPES).nullable().optional();
const bulkActionInputSchema = strapiUtils.z.object({
    documentIds: strapiUtils.z.array(zod.strapiID).min(1)
});
const generateUIDInputSchema = strapiUtils.z.object({
    contentTypeUID: strapiUtils.z.string(),
    field: strapiUtils.z.string(),
    data: strapiUtils.z.looseObject({})
});
const createCheckUIDAvailabilityInputSchema = (regex)=>{
    const pattern = regex ? new RegExp(regex) : /^[A-Za-z0-9-_.~]*$/;
    return strapiUtils.z.object({
        contentTypeUID: strapiUtils.z.string(),
        field: strapiUtils.z.string(),
        value: strapiUtils.z.string().refine((value)=>value === '' || pattern.test(value), {
            error: `Must match the custom regex or the default one "/^[A-Za-z0-9-_.~]*$/"`
        })
    });
};
const validateUIDField = (contentTypeUID, field)=>{
    const model = strapi.contentTypes[contentTypeUID];
    if (!model) {
        throw new ValidationError('ContentType not found');
    }
    if (!_.has(model, [
        'attributes',
        field
    ]) || _.get(model, [
        'attributes',
        field,
        'type'
    ]) !== 'uid') {
        throw new ValidationError(`${field} must be a valid \`uid\` attribute`);
    }
};
const validateKind = zod.validateZodAsync(kindSchema);
const validateBulkActionInput = zod.validateZodAsync(bulkActionInputSchema);
const validateGenerateUIDInput = zod.validateZodAsync(generateUIDInputSchema);
const validateCheckUIDAvailabilityInput = (body)=>{
    let regex;
    const contentType = body.contentTypeUID in strapi.contentTypes ? strapi.contentTypes[body.contentTypeUID] : null;
    if (contentType?.attributes[body.field] && `regex` in contentType.attributes[body.field] && contentType.attributes[body.field].regex) {
        regex = (contentType?.attributes[body.field]).regex;
    }
    const schema = createCheckUIDAvailabilityInputSchema(regex);
    return zod.validateZodAsync(schema)(body);
};

exports.validateBulkActionInput = validateBulkActionInput;
exports.validateCheckUIDAvailabilityInput = validateCheckUIDAvailabilityInput;
exports.validateGenerateUIDInput = validateGenerateUIDInput;
exports.validateKind = validateKind;
exports.validateUIDField = validateUIDField;
