// @strapi/content-manager/dist/server/controllers/validation/relations.js
'use strict';

var strapiUtils = require('@strapi/utils');
var zod = require('../../validation/zod.js');

const validateFindAvailableSchema = strapiUtils.z.object({
    component: strapiUtils.z.string().optional(),
    id: zod.strapiID.optional(),
    _q: strapiUtils.z.string().optional(),
    idsToOmit: strapiUtils.z.array(zod.strapiID).optional(),
    idsToInclude: strapiUtils.z.array(zod.strapiID).optional(),
    page: strapiUtils.z.coerce.number().int().min(1).optional(),
    pageSize: strapiUtils.z.coerce.number().int().min(1).max(100).optional(),
    locale: strapiUtils.z.string().nullable().optional(),
    status: strapiUtils.z.enum([
        'published',
        'draft'
    ]).nullable().optional()
});
const validateFindExistingSchema = strapiUtils.z.object({
    page: strapiUtils.z.coerce.number().int().min(1).optional(),
    pageSize: strapiUtils.z.coerce.number().int().min(1).max(100).optional(),
    locale: strapiUtils.z.string().nullable().optional(),
    status: strapiUtils.z.enum([
        'published',
        'draft'
    ]).nullable().optional()
});
const validateFindAvailable = zod.validateZodAsync(validateFindAvailableSchema);
const validateFindExisting = zod.validateZodAsync(validateFindExistingSchema);

exports.validateFindAvailable = validateFindAvailable;
exports.validateFindExisting = validateFindExisting;
