// @strapi/admin/dist/admin/admin/src/pages/Settings/components/Tokens/FormApiTokenContainer.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import 'react';
import { Box, Flex, Typography, Grid, Field, TextInput } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { useAuth } from '../../../../features/Auth.mjs';
import { LifeSpanInput } from './LifeSpanInput.mjs';
import { TokenDescription } from './TokenDescription.mjs';
import { TokenName } from './TokenName.mjs';
import { TokenTypeSelect } from './TokenTypeSelect.mjs';

const formatAdminUserName = (owner)=>{
    const full = [
        owner.firstname,
        owner.lastname
    ].filter(Boolean).join(' ');
    return full || owner.username || owner.email || '';
};
const FormApiTokenContainer = ({ errors = {}, onChange, canEditInputs, isCreating, values = {}, apiToken = {}, kind, onDispatch, setHasChangedPermissions })=>{
    const { formatMessage } = useIntl();
    const currentUser = useAuth('FormApiTokenContainer', (state)=>state.user);
    const ownerToDisplay = (()=>{
        const owner = kind === 'admin' ? apiToken?.adminUserOwner : undefined;
        if (owner === undefined || owner === null) return null;
        if (typeof owner !== 'object') return null;
        if (currentUser?.id !== undefined && owner.id === currentUser.id) return null;
        return owner;
    })();
    const handleChangeSelectApiTokenType = ({ target: { value } })=>{
        setHasChangedPermissions(false);
        if (value === 'full-access') {
            onDispatch({
                type: 'SELECT_ALL_ACTIONS'
            });
        }
        if (value === 'read-only') {
            onDispatch({
                type: 'ON_CHANGE_READ_ONLY'
            });
        }
    };
    const typeOptions = [
        {
            value: 'read-only',
            label: {
                id: 'Settings.tokens.types.read-only',
                defaultMessage: 'Read-only'
            }
        },
        {
            value: 'full-access',
            label: {
                id: 'Settings.tokens.types.full-access',
                defaultMessage: 'Full access'
            }
        },
        {
            value: 'custom',
            label: {
                id: 'Settings.tokens.types.custom',
                defaultMessage: 'Custom'
            }
        }
    ];
    return /*#__PURE__*/ jsx(Box, {
        background: "neutral0",
        hasRadius: true,
        shadow: "filterShadow",
        paddingTop: 6,
        paddingBottom: 6,
        paddingLeft: 7,
        paddingRight: 7,
        children: /*#__PURE__*/ jsxs(Flex, {
            direction: "column",
            alignItems: "stretch",
            gap: 4,
            children: [
                /*#__PURE__*/ jsx(Typography, {
                    variant: "delta",
                    tag: "h2",
                    children: formatMessage({
                        id: 'global.details',
                        defaultMessage: 'Details'
                    })
                }),
                /*#__PURE__*/ jsxs(Grid.Root, {
                    gap: 5,
                    children: [
                        /*#__PURE__*/ jsx(Grid.Item, {
                            m: 6,
                            xs: 12,
                            direction: "column",
                            alignItems: "stretch",
                            children: /*#__PURE__*/ jsx(TokenName, {
                                error: errors['name'],
                                value: values['name'],
                                canEditInputs: canEditInputs,
                                onChange: onChange
                            })
                        }, "name"),
                        /*#__PURE__*/ jsx(Grid.Item, {
                            m: 6,
                            xs: 12,
                            direction: "column",
                            alignItems: "stretch",
                            children: /*#__PURE__*/ jsx(TokenDescription, {
                                error: errors['description'],
                                value: values['description'],
                                canEditInputs: canEditInputs,
                                onChange: onChange
                            })
                        }, "description"),
                        /*#__PURE__*/ jsx(Grid.Item, {
                            m: 6,
                            xs: 12,
                            direction: "column",
                            alignItems: "stretch",
                            children: /*#__PURE__*/ jsx(LifeSpanInput, {
                                isCreating: isCreating,
                                error: errors['lifespan'],
                                value: values['lifespan'],
                                onChange: onChange,
                                token: apiToken
                            })
                        }, "lifespan"),
                        kind === 'content-api' && /*#__PURE__*/ jsx(Grid.Item, {
                            m: 6,
                            xs: 12,
                            direction: "column",
                            alignItems: "stretch",
                            children: /*#__PURE__*/ jsx(TokenTypeSelect, {
                                value: values['type'],
                                error: errors['type'],
                                label: {
                                    id: 'Settings.tokens.form.type',
                                    defaultMessage: 'Token type'
                                },
                                onChange: (value)=>{
                                    handleChangeSelectApiTokenType({
                                        target: {
                                            value: String(value)
                                        }
                                    });
                                    onChange({
                                        target: {
                                            name: 'type',
                                            value: String(value)
                                        }
                                    });
                                },
                                options: typeOptions,
                                canEditInputs: canEditInputs
                            })
                        }, "type"),
                        ownerToDisplay !== null && /*#__PURE__*/ jsx(Grid.Item, {
                            m: 6,
                            xs: 12,
                            direction: "column",
                            alignItems: "stretch",
                            children: /*#__PURE__*/ jsxs(Field.Root, {
                                name: "adminUserOwner",
                                children: [
                                    /*#__PURE__*/ jsx(Field.Label, {
                                        children: formatMessage({
                                            id: 'Settings.tokens.form.owner',
                                            defaultMessage: 'Owner'
                                        })
                                    }),
                                    /*#__PURE__*/ jsx(TextInput, {
                                        type: "text",
                                        value: formatAdminUserName(ownerToDisplay),
                                        disabled: true,
                                        onChange: ()=>{}
                                    })
                                ]
                            })
                        }, "owner")
                    ]
                })
            ]
        })
    });
};

export { FormApiTokenContainer };
